{
 "metadata": {
  "name": "phylogeny-reconstruction"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Phylogenetic reconstruction"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "What: \n",
      "\n",
      "The process of inferring the evolutionary relationships between organisms or groups of organisms, and usually representing that information as a tree. \n",
      "\n",
      "Why:\n",
      "\n",
      "1. Make functional inferences about genes. \n",
      "2. Understand relationships between organisms (and therefore possible similarities between them). \n",
      "3. Compare the composition of communities of organisms (but we'll come back to this).\n",
      "\n",
      "How:\n",
      "\n",
      "By comparing traits. In our case, traits are columns in a multiple sequence alignment."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Some terminology"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Rooted trees: includes an assumption about the last common ancestor of all sequences.\n",
      "\n",
      "<img src=\"files/images/basic-rooted-tree1.jpg\" width=75%>\n",
      "\n",
      "Unrooted trees: no assumption about the last common ancestor of all sequences\n",
      "\n",
      "<img src=\"files/images/basic-unrooted-tree1.jpg\" width=75%>"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Terminal nodes, tips or leaves** extant organisms, frequently called operational taxonomic units or OTUs. OTUs are families of related organisms. \n",
      "\n",
      "**Internal nodes** hypothetical ancestors - we postulate their existence but often don\u2019t have direct evidence. \n",
      "\n",
      "**Branches** representative of the distance between the nodes.\n",
      "\n",
      "<img src=\"files/images/tree-schematic1.png\">\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We expect that as time progresses, sequences will diverge such that more similar sequences have diverged more recently. The problem of phylogenetic reconstruction however is that we only have the tips. We don't have sequences for the internal nodes, and so we use modern sequences to develop a hypothesis about the evolutionary history of a sequence (and hopefully of the organisms who encode those sequences in their genomes).\n",
      "\n",
      "<img src=\"files/images/sequence-evo-tree.png\">"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**This is very rough - just some code that I pull from a previous year's class, but no annotation yet.**"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# import some functions from python's random module - these will\n",
      "# be used in the modeling process\n",
      "from random import choice, randint\n",
      "# import some math functions from the numpy library (note that this\n",
      "# isn't part of the python standard library)\n",
      "from numpy import log10, average\n",
      "# import argv from the sys module to support basic command line\n",
      "# control of this script\n",
      "from sys import argv\n",
      "\n",
      "#####\n",
      "# Start of function definitions\n",
      "#####\n",
      "\n",
      "def count_differences(sequence1,sequence2):\n",
      "     \"\"\"Count the number of differences between two sequences of the same length\n",
      "     \"\"\"\n",
      "     # confirm that the two sequences are the same length and throw\n",
      "     # an error if they aren't\n",
      "     assert len(sequence1) == len(sequence2), \"Sequences differ in length\"\n",
      "     # initiate a counter for the number of differences\n",
      "     result = 0\n",
      "     # iterate over the two sequences and count the number of\n",
      "     # positions which are not identical\n",
      "     for base1,base2 in zip(sequence1,sequence2):\n",
      "         if base1 != base2:\n",
      "             # this is a commonly used shortcut for incrementing a count\n",
      "             # and is equivalent to the following statement\n",
      "             # result = result + 1\n",
      "             result += 1\n",
      "     return result\n",
      "\n",
      "def evolve_seq(sequence,\n",
      "               substitution_probability=0.01,\n",
      "               mutation_choices=['A','C','G','T']):\n",
      "    \"\"\"Return two child sequences simulating point mutations\n",
      "\n",
      "       An error occurs with probability substitution_probability\n",
      "        independently at each position of each child sequence.\n",
      "    \"\"\"\n",
      "    # Generate two lists for storing the resulting sequences\n",
      "    r1 = []\n",
      "    r2 = []\n",
      "\n",
      "    range_length = 10 ** (-1 * log10(substitution_probability))\n",
      "\n",
      "    for base in sequence:\n",
      "        if randint(0,range_length) == 0:\n",
      "            # a point mutation will occur at this position\n",
      "            # what's wrong with the following statement?\n",
      "            r1.append(choice(mutation_choices))\n",
      "        else:\n",
      "            # no point mutation at this position\n",
      "            r1.append(base)\n",
      "        if randint(0,range_length) == 0:\n",
      "            # a point mutation will occur at this position\n",
      "            # what's wrong with the following statement?\n",
      "            r2.append(choice(mutation_choices))\n",
      "        else:\n",
      "            # no point mutation at this position\n",
      "            r2.append(base)\n",
      "    # convert the lists to strings and return them\n",
      "    return ''.join(r1), ''.join(r2)\n",
      "\n",
      "def main(root_sequence,generations,verbose=False):\n",
      "    # initial some values and perform some basic error checking\n",
      "    assert generations > 0, \"Must simulate one or more generations.\"\n",
      "    # can you simplify the following test?\n",
      "    for base in root_sequence:\n",
      "        assert base != 'A' or base != 'C' or base != 'G' or base != 'T',\\\n",
      "         \"Invalid base identified: %s. Only A, C, G, or T are allowed.\" % base\n",
      "    # initialize a list of the previous generations sequences - this gets used\n",
      "    # in the for loop below. since we'll start with the first generation of\n",
      "    # children, root_sequence is the previous generation's sequence\n",
      "    previous_generation_sequences = [root_sequence]\n",
      "\n",
      "    # iterate over each generation (why do we add one to generations?)\n",
      "    for i in range(1,generations+1):\n",
      "        # print the generation number and the current number of sequences\n",
      "        print \"Generation: %d (Number of child sequences: %d)\" % (i,2**i)\n",
      "        # create a list to store the current generation of sequences\n",
      "        current_generation_sequences = []\n",
      "        # create a list to store the differences in each current generation\n",
      "        # sequence from the root sequence\n",
      "        difference_counts = []\n",
      "        # iterate over the sequences of the previous generation\n",
      "        for parent_sequence in previous_generation_sequences:\n",
      "            # evolve two child sequences\n",
      "            r1, r2 = evolve_seq(parent_sequence)\n",
      "            # count the differences in the first sequence (from root_sequence)\n",
      "            r1_diffs = count_differences(root_sequence,r1)\n",
      "            # append the count of differences to the list of difference counts\n",
      "            difference_counts.append(r1_diffs)\n",
      "            # add the new sequence to the list of this generation's sequences\n",
      "            current_generation_sequences.append(r1)\n",
      "            # count the differences in the second sequence (from root_sequence)\n",
      "            r2_diffs = count_differences(root_sequence,r2)\n",
      "            # append the count of differences to the list of difference counts\n",
      "            difference_counts.append(r2_diffs)\n",
      "            # add the new sequence to the list of this generation's sequences\n",
      "            current_generation_sequences.append(r2)\n",
      "            if verbose:\n",
      "                # if the caller specified verbose output, print the actual sequences\n",
      "                print \"  %s %d\" % (r1, r1_diffs)\n",
      "                print \"  %s %d\" % (r2, r2_diffs)\n",
      "        # print summary information: the average number of differences in the current\n",
      "        # generation from root_sequence\n",
      "        print \"Mean differences %1.3f\\n\" % average(difference_counts)\n",
      "        # current_generation_sequences becomes the next generation's\n",
      "        # previous_generation_sequences\n",
      "        previous_generation_sequences = current_generation_sequences\n",
      "\n",
      "    # upon completion of all generations, return the last generation's sequences\n",
      "    return previous_generation_sequences\n",
      "\n",
      "#####\n",
      "# End of function definitions\n",
      "#####"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sequence_length = 100\n",
      "num_generations = 5\n",
      "\n",
      "# generate a random sequence composed of ['A', 'C', 'G', 'T']\n",
      "# of length sequence_length\n",
      "root_sequence = []\n",
      "for i in range(int(sequence_length)):\n",
      "    root_sequence.append(choice(list('ACGT')))\n",
      "root_sequence = ''.join(root_sequence)\n",
      "\n",
      "# run the simulation and get the final generation of sequences\n",
      "sequences = main(root_sequence,int(num_generations))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Generation: 1 (Number of child sequences: 2)\n",
        "Mean differences 0.500\n",
        "\n",
        "Generation: 2 (Number of child sequences: 4)\n",
        "Mean differences 1.000\n",
        "\n",
        "Generation: 3 (Number of child sequences: 8)\n",
        "Mean differences 1.500\n",
        "\n",
        "Generation: 4 (Number of child sequences: 16)\n",
        "Mean differences 2.188\n",
        "\n",
        "Generation: 5 (Number of child sequences: 32)\n",
        "Mean differences 2.594\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# print the final generation of sequences in fasta format\n",
      "for i,s in enumerate(sequences):\n",
      "    print '>seq%d\\n%s' % (i,s)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">seq0\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTATTGTTAAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq1\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGATAAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq2\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq3\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq4\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTAATAA\n",
        ">seq5\n",
        "CAGCTGGATCAGATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTAATAA\n",
        ">seq6\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTTTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq7\n",
        "CAGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTCAGCAGGTATGCTGGACTATGCTATTAA\n",
        ">seq8\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGGTATAGAACAACAACCTCTCACACGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACCATGCTATGAA\n",
        ">seq9\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGGTATAGAACAACAACCTCTCACACGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACCATGCTATGAA\n",
        ">seq10\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACCATGCTATGAA\n",
        ">seq11\n",
        "CGGCTGGATCACATGGAAACAACCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACCATGCTATGAA\n",
        ">seq12\n",
        "CGGCTGGATCACATGGAAACAAGCCCTGATATAGAACAACACCATCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATGAA\n",
        ">seq13\n",
        "CGGCTGGATCACATGGAAACAAGCCCTGATATAGAACAACACCCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATGAA\n",
        ">seq14\n",
        "CGGATGGATCACATGGAAACAAGCCGTGATATAGAACAACACCCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATGAA\n",
        ">seq15\n",
        "CGGCTGGTTCACATGGAAACAAGCCGTGATATAGAACAACACCCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGCTATGAC\n",
        ">seq16\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq17\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq18\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq19\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq20\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq21\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq22\n",
        "CGGCTGGCTCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq23\n",
        "CGGCTGGCTCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq24\n",
        "CGGCTGGATCACTTGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTTTTAAGGGCTCTTGTTAAGCAGGTATGCTGGAGTATGGTATGAA\n",
        ">seq25\n",
        "CGGCTGGATCACTTGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTTTTAAGGGCTCTTGTTAAGCAGGTATGCTGGAGTATGGTATGAA\n",
        ">seq26\n",
        "CGGCTGGATCATATGGAAACAAGCCGTGATATAGGACAACAACCTCTCACATGTTTTTTAAGGGCTCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq27\n",
        "CGGCTGGATCATATGGAAACAAGCCGTGATATAGGACAACAACCTCTCACATGTTTTTTAAGGGCCCTTGTTAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq28\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATTGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGATAAGCAGGTATGCTGGACTATGGTATGGA\n",
        ">seq29\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTCTCACATGTTTATTAAGGGCTCTTGATAAGCAGGTATGCTGGACTATGGTATGGA\n",
        ">seq30\n",
        "CGGCTGGATCACATGGAAGCAAGCCGTGATATAGAACAACAACCTATCACATGTTTATTAAGGGCTCTTGATAAGCAGGTATGCTGGACTATGGTATGAA\n",
        ">seq31\n",
        "CGGCTGGATCACATGGAAACAAGCCGTGATATAGAACAACAACCTATCACATGTTTATTAAGGGCTCTTGATAAGCAGGTATGCTGGACTATGGTATGAA\n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**TODO: Implement UPGMA, build tree from these sequences, visualize tree**"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}