{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence mapping and clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A common need in bioinformatics is, given a number of *query* sequences, group them based on their similarity either to one another, or their similarity to sequences in an external reference database, or to both. The most common way to do this is using sequence alignment (you may be noticing a theme here...).\n",
    "\n",
    "The process of searching a set of sequences against a reference database to find their best match is typically referred to a **sequence mapping**. One example of this would be in *genome re-sequencing*. If you're searching for polymorphisms in the human genome that may be associated with a phenotype (e.g., a particular disease) you might begin by sequencing the full human genome. Because the human genome has been fully sequenced and (mostly) assembled, you could map your short sequence reads against the full human genome, and then search for loci (or co-located sets of one or more bases) that vary with the phenotype of interest. Because this process is generally performed with DNA sequencing reads, you may also hear it referred to as *read mapping*.\n",
    "\n",
    "A similar process can be applied if there is no reference database to search against. In this case, sequences will be grouped together based on their similarity to one another. This is most often applied to reads of a single gene or locus across the genomes of many different organisms. This process is referred to as **de novo sequence clustering**, and one field where this is common is **microbiomics**, or the study of whole communities of microorganisms. Because we don't know how to culture the vast majority of microbes, most of what we know about the composition of microbial communities (e.g., in free-living environments, such as the ocean, soil, or surfaces in our homes or offices, or in host-associated environments, such as the human gut) is based on sequencing specific marker genes such as the 16S rRNA from all community members. If we obtain a large number of sequence reads, many of the things we want to do with them (such as identify their taxonomic origin, or understand where they fall in a phylogenetic tree) is too computationally intensive to achieve. So instead, we group sequences that are identicial or highly similar in composition into **Operational Taxonomic Units (OTUs)**, and we choose a single representative of that OTU to work with downstream. For example, if we have a group of 16S rRNA reads that are within 97% identity to one member of that cluster (the cluster centroid) we may assume that the taxonomic origin of the cluster centroid is the same as the taxonomic origin of all of the sequences in the group. This is an *assumption* - it may or may not be true - but it is a necessary evil given the current technology. \n",
    "\n",
    "Another application of grouping similar sequences (or **OTU clustering**, or **OTU picking**, as it is sometimes referred to) is in grouping sequences in a database before investigating them, to reduce taxonomic bias in the database. For example, *E. coli* is one of the most heavily sequenced microbes. If you're interested in understanding the frequency of variants of a specific gene across a range of microbial diversity, you might begin by obtatining all sequences of that gene from [GenBank](https://www.ncbi.nlm.nih.gov/genbank/). Because there may be many more *E. coli* sequences, purely because of sequencing bias, you'd likely want to group your sequences into OTUs before computing variant frequencies, so your calculations are not biased toward the frequencies in *E. coli*, as hundreds of E. coli sequences would likely group to one or a few closely related OTUs. In other words, you're trying to find a **divergent set** of sequences to work with (and an [aptly named tool](http://www.ncbi.nlm.nih.gov/pubmed/16769708) was published in 2006 to automate this process).\n",
    "\n",
    "We have learned the key tools we need for both sequence mapping and clustering in previous chapters. **Because the process of read mapping is nearly identical to database searching, in this chapter we'll start by exploring how to perform de novo sequence clustering**. At the end of the chapter we'll look at a case where we combine sequence clustering with sequence mapping, which arose to deal with massive sequence datasets generated in microbiomics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "\n",
    "from __future__ import print_function, division\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.core import page\n",
    "page.page = print\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "from skbio import BiologicalSequence, DNA, Alignment\n",
    "from skbio.alignment import SequenceCollection\n",
    "from skbio.parse.sequences import parse_fasta\n",
    "from skbio.alignment import local_pairwise_align_ssw, global_pairwise_align_nucleotide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *De novo* clustering of sequences by similarity "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The algorithm at the core of *de novo* clustering is sequence alignment. In an ideal world, we would perform a full multiple sequence alignment of all of our sequences, compute their pairwise similarities (or dissimilarities), and use those values to group sequences that are above some *similarity threshold* into *OTU clusters* (just *OTUs* from here). As we discussed in the mutliple sequence alignment chapter however, that is infeasible for more than a few tens of sequences due to computational and memory requirements. Even progressive alignment can't typically handle more than a few tens of thousands of sequences (at least with the currently available implementations, that I am aware of), so OTU clustering is generally acheived by picking pairs of sequences to align. You'll notice in this section that many of the heuristics that have been applied for speeding up database searching are similar to the heuristics applied for OTU clustering.\n",
    "\n",
    "We'll work with SSW for Smith-Waterman pairwise alignment with affine gap scoring here, though in principle any pairwise aligner could be substituted. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on built-in function local_pairwise_align_ssw in module skbio.alignment._ssw_wrapper:\n",
      "\n",
      "local_pairwise_align_ssw(...)\n",
      "    Align query and target sequences with Striped Smith-Waterman.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    sequence1 : str or BiologicalSequence\n",
      "        The first unaligned sequence\n",
      "    sequence2 : str or BiologicalSequence\n",
      "        The second unaligned sequence\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    ``skbio.alignment.Alignment``\n",
      "        The resulting alignment as an Alignment object\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    This is a wrapper for the SSW package [1]_.\n",
      "    \n",
      "    For a complete list of optional keyword-arguments that can be provided,\n",
      "    see ``skbio.alignment.StripedSmithWaterman``.\n",
      "    \n",
      "    The following kwargs will not have any effect: `suppress_sequences` and\n",
      "    `zero_index`\n",
      "    \n",
      "    If an alignment does not meet a provided filter, `None` will be returned.\n",
      "    \n",
      "    References\n",
      "    ----------\n",
      "    .. [1] Zhao, Mengyao, Wan-Ping Lee, Erik P. Garrison, & Gabor T.\n",
      "       Marth. \"SSW Library: An SIMD Smith-Waterman C/C++ Library for\n",
      "       Applications\". PLOS ONE (2013). Web. 11 July 2014.\n",
      "       http://www.plosone.org/article/info:doi/10.1371/journal.pone.0082138\n",
      "     \n",
      "    See Also\n",
      "    --------\n",
      "    skbio.alignment.StripedSmithWaterman\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(local_pairwise_align_ssw)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"float: right; margin-left: 30px;\"><img style=\"float: right; margin-left: 30px;\" src=\"files/images/cluster-legend.png\" align=right></div>\n",
    "\n",
    "In the figures that follow, points represent sequences. A line projecting from a point indicates a dissimilarity range from that sequence (e.g., 10% dissimilar). The circle defined by the point (sequence) and the line (dissimilarity range) define the space containing all sequences within a given dissimilarity threshold.\n",
    "\n",
    "These illustrations are used to describe OTUs. When a circle or set of overlapping circles are filled with grey, that defines the space that a new sequence can fall in if it will be considered part of the OTU. In the the following sections we'll explore different ways that this can be defined.\n",
    "\n",
    "*These figures attempt to illustrate several ideas about OTU clustering, but because the space that we're working in is two dimensional, we can't perfectly represent the process. These figures are useful as visual aid, but should not be considered mathematically robust.* \n",
    "\n",
    "\n",
    "Let's define a collection of sequences to work with. These are derived from the [Greengenes](http://greengenes.secondgenome.com/) [13_8](ftp://greengenes.microbio.me/greengenes_release/gg_13_5/) database, and we're pulling them from the [QIIME default reference project](https://github.com/biocore/qiime-default-reference). We can load these as a list of sequences using ``skbio.parse.sequences.parse_fasta``, and count them by taking the length of the list. For the sake of runtime, we'll work with only a small random subset these sequences.\n",
    "\n",
    "**Our goal here will be to group these sequences into OTUs based on some similarity threshold that we define.** If we set this similarity threshold at 90%, meaning that the sequences within that OTU are 90% identicial (either to each other, or maybe to some representative of that cluster - we'll explore some variants on that definition below), we'd call these *90% OTUs*.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99\n"
     ]
    }
   ],
   "source": [
    "from qiime_default_reference import get_reference_sequences \n",
    "from random import random\n",
    "\n",
    "seqs_16s = []\n",
    "fraction_to_keep = 0.001\n",
    "for e in list(parse_fasta(get_reference_sequences())):\n",
    "    if random() < fraction_to_keep:\n",
    "        seqs_16s.append(BiologicalSequence(e[1], id=e[0]))\n",
    "seqs_16s = SequenceCollection(seqs_16s)\n",
    "print(seqs_16s.sequence_count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def show_clusters(clusters, print_clusters=True, plot_clusters=True, plot_labels=False):\n",
    "    G = nx.Graph()\n",
    "    for c in clusters:\n",
    "        G = nx.union(G, c)\n",
    "        if print_clusters: print(\"%s: %s\" % (c.graph['id'], [s.id for s in c.nodes()]))\n",
    "    if plot_clusters:\n",
    "        pos=nx.spring_layout(G)\n",
    "        nx.draw_networkx_nodes(G, pos)\n",
    "        if G.number_of_edges() > 0:\n",
    "            nx.draw_networkx_edges(G, pos)\n",
    "        if plot_labels: nx.draw_networkx_labels(G, pos, labels={d:d.id for d in G.nodes()})\n",
    "        _ = plt.axis('off')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Furthest neighbor clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first approach we'll look at is one that has been called *furthest neighbor*, because whether a sequence becomes a member of a cluster is defined by it's most dissimilar (furthest) \"neighbor\" in that cluster.\n",
    "\n",
    "The way this algorithm works is that we start with our list of sequences. Because this is *de novo* clustering, by definition our first sequence is added to a new cluster (because there are no pre-existing clusters). We'll call this `OTU 1`. We then iterate over the remaining sequences. For the second sequence we compute its pairwise alignment with the first, followed by computing their percent similarity. If their percent similarity is greater than or equal to the similarity threshold, we add the second sequence to `OTU 1`. If is it less than the similarity threshold, we create a new OTU, `OTU 2`, and add the second sequence to that cluster. \n",
    "\n",
    "We continue iterating over the remaining sequences. For each sequence, we compute the pairwise similarity between that sequence and all sequences in each OTU. **If the percent similarity between a query sequence and *all* sequences in a given OTU (say OTU $x$ is greater than the similarity threshold, we add the sequence to OTU $x$.)** Otherwise we check the next OTU in the list. If this criteria is not met for any of the OTUs, then we define a new OTU to add the sequence to. \n",
    "\n",
    "Here's what that process would look like for these six sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = DNA('AAAAAAAAAA', 's1')\n",
    "s2 = DNA('AAAAATTTTT', 's2')\n",
    "s3 = DNA('AAAAAAACCA', 's3')\n",
    "s4 = DNA('CCCCAATTTT', 's4')\n",
    "s5 = DNA('ACCAAATTTT', 's5')\n",
    "s6 = DNA('AGGAAAAAAA', 's6')\n",
    "\n",
    "aln1 = Alignment([s1, s2, s3, s4, s5, s6])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our first sequence, ``s1``, will define a new OTU. We'll call that OTU ``OTU 1``. We'll great a list of OTUs and add this as our first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "otus = []\n",
    "G=nx.Graph(id=\"OTU 1\")\n",
    "G.add_node(s1)\n",
    "otus.append(G)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our second sequence, ``s2`` falls outside of the similarity threshold to ``S1``, it will also define a new OTU. We'll call that OTU ``OTU 2``."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5\n"
     ]
    }
   ],
   "source": [
    "print(s2.fraction_same(s1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['s1']\n",
      "OTU 2: ['s2']\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEACAYAAACznAEdAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAACMZJREFUeJzt3U2IXtdhgOH3KtZIGqwfR4j81osQQhoX4kTZlSxDEkJC\n",
       "0cYVhUKzULMNFYgULbwwDQZBkq2hhmbnjSk2tO6i3iRZJDBBgiiQ0EWrTSCKjPWTkUYKul18Z7Ds\n",
       "2JJsaTTDzPOsdL+5V5zV93LPufd80zzPAcCuzR4AAFuDIABQCQIAgyAAUAkCAIMgAFAJAgCDIABQ\n",
       "CQIAgyAAUAkCAIMgAFAJAgCDIABQCQIAgyAAUAkCAIMgAFAJAgCDIABQCQIAgyAAUAkCAIMgAFAJ\n",
       "AgCDIABQCQIAgyAAUAkCAIMgAFAJAgCDIABQCQIAgyAAUAkCAIMgAFAJAgCDIABQCQIAgyAAUAkC\n",
       "AIMgAFAJAgCDIABQCQIAgyAAUAkCAMNjmz2A7WqapoPV4XF4aZ7ny5s5HoB7cYfwEE3TtGeapuNP\n",
       "TNPZpbp4pM4dqXNLdfGJaTo7TdPxaZqWNnucAO9mmud5s8ewLeyapmf21gtfqOlk7f9mb91+3ape\n",
       "rc7U1bM136gTt+f5pc0bLcCfE4SHYN80fffxeu61Wj56j3NXqq/V6rU6fX2ef/AoxgdwPwThAe2a\n",
       "pmcO14srtfzkfV5zoTpaq5fq2+4UgK1CEB7ANE179tXvf1oHvvge5/yq+qfql9Wl6vb4fKX6cl25\n",
       "Xkfmeb75CIYLcFcWlR/Msadreq8YVC1Vf1v96zs+P1p9vqbq2EYNDuD9EIQHcKhOnaz968fPV5+s\n",
       "DlSfrV6vPlP9Q/W5d7n+ZO0/VKcexVgB7sWU0Qc0TdPBpbr4x9r9WPWb6ivVL6qPtlgn+FP1qXH+\n",
       "/7SIw+07/o9b1eN16+Zi2sh7CsCmcofwwR0+WGvrj5Z+qFqrzrf4on+yt2LwXnZXB+pm9eENGyXA\n",
       "fRKEh+TT1Q+rZ6uPVMer323mgADeJ0H44C5drj237vjgePWT6v9arBbfa3HgVnVlse78xsYMEeD+\n",
       "CcIHNM/z5eX69avj+LctFpHXqj3V3hbTSFU3WswLNf6+Nv79SrVc560fAFuBIDyAN+v5M3W1Fl/y\n",
       "36uOVB+r/lB9v/rfarn6qxZ3DfuqvxzXn6mrby4eTgLYdJ4yegD382Lae/FiGrDVuEN4APM8r92o\n",
       "E1+t1Qvv47oLLfYzulEnxADYKgThAd2e55eu1emjtbpyH+evtNjH6Fqdto8RsJWYMnpI1re/fnps\n",
       "f/2t3r799Sst1gzO2f4a2KIE4SEaP35z7FCdWq2nxktnXaml5To/FpBfNk0EbEWCsEHGT2iuv4H8\n",
       "hkdLga1OEACoLCoDMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgC\n",
       "AJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIA\n",
       "lSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCV\n",
       "IAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUg\n",
       "ADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAA\n",
       "MAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAw\n",
       "CAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAI\n",
       "AgCVIAAwCAIAlSAAMAgCAJUgADAIAgCVIAAwCAIAlSAAMAgCAJUgADAIAgBVPbbZAwDYSqZpOlgd\n",
       "HoeX5nm+vJnjeZTcIQA73jRNe6ZpOv7ENJ1dqotH6tyROrdUF5+YprPTNB2fpmlps8e50aZ5njd7\n",
       "DACbZtc0PbO3XvhCTSdr/zd7a+rkVvVqdaaunq35Rp24Pc8vbd5oN5YgADvWvmn67uP13Gu1fPQe\n",
       "565UX6vVa3X6+jz/4FGM71ETBGBH2jVNzxyuF1dq+cn7vOZCdbRWL9W3t+OdgjUEYMeZpmnP3nrh\n",
       "v+4Sg3+rvlQdrP6iOlV9onqtlvfWC9txTUEQgJ3o2NM1ffEuJ1yvflRdqn5e/Xd1pjpafb6m6tiG\n",
       "j/IREwRgxzlUp07W/vXj56tPVgeqz1avV9+p/rrFAvPHq7+rfjbOP1n7Dy1uGrYVawjAjjJN08Gl\n",
       "uvjH2v1Y9ZvqK9Uvqo+2WCf4U/Wpd1z3N9Xnqn9p8fTR43XrZh3ZTu8puEMAdprDB2tt/dHSD1Vr\n",
       "1fkWX/RP9ucxeLH6ZXVyHO+uDtTN6sMbP9xHRxCAHe3T1Q+rZ6uPVMer393x93+v/rn6z7bZt/+7\n",
       "MGUE7CjrU0bXavfud/ztavWPLdYNfly9Vv199R8tnjhaZ8oIYBuY5/nycv361XH82xaLyGvVnmpv\n",
       "i2mk11ssJL/c22NQ9Uq1XOe3UwxKEIAd6M16/szihqC16nvVkepj1R9aLBw/1+KEr7d4HGl/9Y1x\n",
       "/Zm6+ubi4aRtxZQRsONM07RnX/3+p3Xgbu8ivJuV6st15fpiuujmRoxvs7hDAHaceZ7XbtSJr9bq\n",
       "hfdx3YUW+xndqBPbLQYlCMAOdXueX7pWp4/W6sp9nL/SYh+ja3V6O+5jVKaMgB1uffvrp8f219/q\n",
       "7dtfv9JizeCc7a8Btr+xUd2xQ3VqtZ4aL511pZaW6/xYQH55O04T3UkQAO4wfkJz/R20N7bbo6V3\n",
       "IwgAVBaVARgEAYBKEAAYBAGAShAAGAQBgEoQABgEAYBKEAAYBAGAShAAGAQBgEoQABgEAYBKEAAY\n",
       "BAGAShAAGAQBgEoQABgEAYBKEAAYBAGAShAAGAQBgEoQABgEAYBKEAAYBAGAShAAGAQBgEoQABgE\n",
       "AYBKEAAYBAGAShAAGAQBgEoQABgEAYBKEAAYBAGAShAAGAQBgKr+HzYdt0wB/JuWAAAAAElFTkSu\n",
       "QmCC\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x100629c90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "G=nx.Graph(id=\"OTU 2\")\n",
    "G.add_node(s2)\n",
    "otus.append(G)\n",
    "show_clusters(otus, plot_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now imagine that our third sequence, ``s3`` falls within the range of ``s1``. We'd cluster ``s3`` into ``OTU 1`` since it's within the similarity range of all of the sequences in ``OTU 1`` (for now that's just ``s1``). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1 0.2\n",
      "OTU 2 0.5\n"
     ]
    }
   ],
   "source": [
    "for i, otu in enumerate(otus, start=1):\n",
    "    print(' '.join([\"OTU %d\" % i] + map(str,[s.fraction_same(s4) for s in otu])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have three sequences clustered into two OTUs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['s1', 's3']\n",
      "OTU 2: ['s2']\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEACAYAAABI5zaHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAADjxJREFUeJzt3V9sled9wPHvA9gGC2OIR5NGC5q6bFpDpUJcacuabEun\n",
       "rmk7WxWbhKzdZRX3WZlQN6RFW6SKDSmZqlUaF9G2i6lcNJrsqCFbmyYL/V9X0A2msF0kaEuqEDNs\n",
       "gwE749nF8569hwyIweeP/TvfjxTBsc/BD4r01ctznt97Us4ZSVIs67q9AElS6xl3SQrIuEtSQMZd\n",
       "kgIy7pIUkHGXpICMuyQFZNwlKSDjLkkBGXdJCsi4S1JAxl2SAjLukhSQcZekgIy7JAVk3CUpIOMu\n",
       "SQEZd0kKyLhLUkDGXZICMu6SFJBxl6SAjLskBWTcJSkg4y5JARl3SQrIuEtSQMZdkgIy7pIUkHGX\n",
       "pICMuyQFZNwlKaAN3V6AJHVKSmkYGKkezuScZ7u5nnbyyl1SaCmlgZTSxLaUTvTDue1wcjuc7Idz\n",
       "21I6kVKaSCn1d3udrZZyzt1egyS1xbqU9m6EI7sh7YehMertiiVgCjgM8ycgX4F913I+2r3VtpZx\n",
       "lxTSppSe2AxPHYPB0fd57jTwGCxchIOXc366E+trN+MuKZx1Ke0dgWenYXDHMl9zFhiFhRl4PMIV\n",
       "vHGXFEpKaWATvH0ctjx4m6+dBh6BucuwPee82I71dYpvqEqKZs8uSLcK+98CHwOGgfuAA8D/AKPA\n",
       "RyEBe9q+yjYz7pJC2QoH9sPQrZ5zGfhLYAb4PvBN4HD1vf0wtLX0fk1zW0ZSGCml4X44dwn6Gqdi\n",
       "DgFfBuaAe4GvAJ94z+ueBr4FTFJO0WyGpcWyNbNmz8E7xCQpkpFhuLoB+gBeA/4K+BFwD+VN03dv\n",
       "8KJXgI9Uv+8DtsDiO3AXYNwlabVZD1wFTlHGUm90cuZZ4MfVr5G45y4pkplZGFiqHtwPPAM8CdwN\n",
       "TABvNT35H4A/Al6gXKZD2ZaZg37gfCcW3C7GXVIYOefZQTg91fS1CeBV4A3KMZjGO6XHgH3A88DO\n",
       "pudPAoNwai3vt4NxlxTMBTh0GOYBzgAvUbZmBoCNlK2al4DfA56jHIlsdhjmL5T3Ydc0T8tICqV5\n",
       "iKkP+Dzwb5Q3Sj8O/DUl7McpwW/4NeBPiTPEZNwlhePtB9yWkRTQtZyPXoSDo7AwvYznT1PCfhEO\n",
       "Rgg7eOUuKbDGLX93Vbf8Hef6W/5OUvbYT3rLX0laW6oP4tizFQ4swM4tsAjluOMgnKrePH1ure+x\n",
       "v5dxl9Qzqo/ZaxxpP7/WjzveinGXpIB8Q1WSAjLukhSQcZekgIy7JAVk3CUpIOMuSQEZd0kKyLhL\n",
       "UkDGXZICMu6SFJBxl6SAjLskBWTcJSkg4y5JARl3SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFZNwl\n",
       "KSDjLkkBGXdJCsi4S1JAxl2SAjLukhSQcZekgIy7JAVk3CUpIOMuSQEZd0kKyLhLUkDGXZICMu6S\n",
       "FJBxl6SAjLskBWTcJSkg4y5JARl3SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFZNwlKSDjLkkBGXdJ\n",
       "Csi4S1JAxl2SAjLukhSQcZekgIy7JAVk3CUpIOMuSQEZd0kKyLhLUkDGXZICMu6SFJBxl6SAjLsk\n",
       "BWTcJSkg4y5JARl3SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFZNwlKSDjLkkBGXdJCsi4S1JAxl2S\n",
       "AjLukhSQcZekgIy7JAVk3CUpIOMuSQEZd0kKyLhLUkDGXZICMu6SFJBxl6SAjLskBWTcJSkg4y5J\n",
       "ARl3SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFtKHbC1gNUkrDwEj1cCbnPNvN9UjSSvXslXtKaSCl\n",
       "NLEtpRP9cG47nNwOJ/vh3LaUTqSUJlJK/d1epyTdiZRz7vYaOm5dSns3wpHdkPbD0Bj1P2GWgCng\n",
       "MMyfgHwF9l3L+Wj3VitJt6/n4r4ppSc2w1PHYHD0fZ47DTwGCxfh4OWcn+7E+iSpFXoq7utS2jsC\n",
       "z07D4I5lvuYsMAoLM/C4V/CS1oqe2XNPKQ1shCMv3iLsXwV+CRgGfgbYQ9muOQaDG+GIe/CS1oqe\n",
       "iTuwZxekB2/xhI8D/wzMAm8Ag8AfAKPARyFRei9Jq17PxH0rHNgPQ43Hh4CfBbZQrtZfAu4DPlB9\n",
       "PwPrgQ9Wj/fD0FY40LkVS9Kd64k995TScD+cuwR9G4DXgE8CPwDuoeyrvwt8CDgO/DYwB/w68CLQ\n",
       "TzlFsxmWFmG75+AlrXa9cuU+MgxXG8cd1wNXgVOUaO+ghB3gYeAC8J9AH/CH1df7gC2wCNzVqUVL\n",
       "0p3qlbhf537gGeBJ4G5gAnjrPc+5F/gz4O86ujJJao1eifvMLAwsNX1hAniV8sZp4sab6UuUN1Ub\n",
       "v58rOzTn27lQSWqFnoh7znl2EE5PVY/PUN5AvQoMABspWzV/T9l/hxL9PwZ+p3o8WZ53CRhLKbk1\n",
       "I2lV64m4A1yAQ4dhHkrUvwhsp5yGeQf4EnCachxyM/AbwEPAn1ev/wu4OAdHgd8FXk8pvZxS+kJK\n",
       "6Rc6+heRpGXoidMyUIaYNsHbx2HLrc6638g08AjMXS4nZRZTSpuATwDjwBjlcM0k5bY03805v9vS\n",
       "xUvSbeqZuEN7bj+QUloHPEgd+vuAr1Ni/2LOeb5Fy5ekZeupuEP7bxyWUtpBOSo/Dvwq8F2qq/qc\n",
       "89lbvVaSWqXn4g71LX93Vbf8Hef6W/5OUm75e3KFt/xNKQ0Bv0UJ/WeA/6Js3UwC0znnayv+y0jS\n",
       "DfRk3AGqm4Dt2QoHFmBnNaDEHPQPwqkL5Q4Fz+WcF1v089ZT3qMdo8R+GHieEvpv5pwvt+LnSBL0\n",
       "cNybVR+z1zjeeL4TtxeoTtmMVf+NAt+iXNU/n3P+abt/vqTYjPsqUJ2b/zQl9J+iHMVvnL75l+z/\n",
       "JEm3ybivMtV20SOUrZtxygBtI/SvtGqbSFJsxn0VSyklYCf1McsPA/9Iif0LOeeZLi5P0ipm3NeQ\n",
       "lNLdwGcpsX8UOEF1+ibnfKaba5O0uhj3NappSrbxpuw89TFLp2SlHmfcA2iakm0cs2xMyU5RpmTn\n",
       "urg8SV1g3ANqmpIdo9wL7XvUU7JvdHNtkjrDuAdXTcl+knJF/1ngTerTNz9ySlaKybj3kGpK9leo\n",
       "T99s4/op2YUuLk9SCxn3HpZSup96n34UeJkSeqdkpTXOuAuAlNI2ypTsOPWUbOP0jVOy0hpj3PX/\n",
       "NE3JNq7q11MiP4lTstKaYNx1S9WU7APU+/QPUKZkp4CvOyUrrU7GXbelaUp2jDJEdZL6mOVr3Vyb\n",
       "pJpx1x2rpmQfpb6qv0h9zPI7TslK3WPc1RLV9k3zlOwO4AXqz5J1SlbqIOOutkgp3Uf9WbKNKdkp\n",
       "yvbN611cmtQTjLvarmlKdowS/Depj1k6JSu1gXFXR1VTsr9M/WEkjSnZKeAbTslKrWHc1VVNU7Jj\n",
       "wMeAV6inZN/q5tqktcy4a9WopmQfo1zRPwb8O/Xpm584JSstn3HXqpRS6uP6z5JdT71P/0rO+WoX\n",
       "lyetesZdq17TlGzjmOUDwD9RQu+UrHQDxl1rTkrpA9SfJduYkm18lqxTshLGXWtcSmkj9WfJjgOX\n",
       "qPfpv+2UrHqVcVcY1fbNburbIfwc13+W7Gz3Vid1lnFXWE1TsmPAw8D3qW9y9noXlya1nXFXT0gp\n",
       "beb6z5L9KfXpmx92Yko2pTQMjFQPZ/yXhNrJuKvnNE3JNvbpR6g/S7alU7IppQFgz1Y4sAAPDMNV\n",
       "gFkYGITTF+AQ8DU/AEWtZtzV81JKP08d+saU7BRlSvbNO/1z16W0dyMc2Q1pPwyNARuq7y1VP+Aw\n",
       "zJ+AfAX2Xcv56Mr+JlLNuEtNmqZkx6pf/4N6+2bZU7KbUnpiMzx1DAZH3+e50+UHLVyEg5dzfnoF\n",
       "y5f+j3GXbqKakn2Yekq2j/qY5cs3m5Jdl9LeEXh2GgZ3LPNnnQVGYWEGHvcKXq1g3KVlqI5Zfpj6\n",
       "mOVO4BvUU7LvVM8b2ARvH4ctD97kz/pX4AvAj4EZoPFO7jTwCMxdhu3uwWuljLt0B6op2c9QYv+b\n",
       "wE8oob/2EPzJd2DoZq89A3yb8i7u56jjDvAQzH8P9uWcv9q2xasnGHdphaop2UeB8SH4/b+Bvj3V\n",
       "9w4BXwbmgHuBr1DGaaFs5v8i18f9a8Dn4cR/57y7E2tXXMZdapGU0nA/nLsEfRuA1ygH638A3EPZ\n",
       "V38X+FD1/BvFfQnYDEuLZWvGc/C6Y+u6vQApkJFhuNo47riecqj9FCXaO6jDfjN9wBZYBO5q2yrV\n",
       "E4y71Cb3A88ATwJ3AxOAHy2lTjHuUuvMzMLAUtMXJoBXgTeABBx4nz9gCZiDfuB8e5aoXmHcpRbJ\n",
       "Oc8Owump6vEZ4CXK1swAsJGyVQNwhbL3QvX9xoH5SWAQTrnfrpUy7lILXYBDh2EeSrC/CGwHPgi8\n",
       "A3wJeB0YBD5CuZrfRDlAD+V2BNX9ZqQV8bSM1ELLGWK6GYeY1EpeuUstlHO+egX2fQoWzt7G685S\n",
       "7i9zpQwwGXatmHGXWuxazkcvwsFRWJhexvOnKfeVuQgHva+MWsVtGalNGrf83VXd8nec62/5O0nZ\n",
       "Yz/pLX/VBsZdaqOUUj/1h3XsrAaUmIP+QThVvXn6nFsxajXjLnVI9TF7jcnT8x53VDsZd0kKyDdU\n",
       "JSkg4y5JARl3SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFZNwlKSDjLkkBGXdJCsi4S1JAxl2SAjLu\n",
       "khSQcZekgIy7JAVk3CUpIOMuSQEZd0kKyLhLUkDGXZICMu6SFJBxl6SAjLskBWTcJSkg4y5JARl3\n",
       "SQrIuEtSQMZdkgIy7pIUkHGXpICMuyQFZNwlKSDjLkkBGXdJCsi4S1JAxl2SAjLukhSQcZekgIy7\n",
       "JAVk3CUpIOMuSQH9L8IyFsXlAQ5UAAAAAElFTkSuQmCC\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10663e8d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "otus[0].add_edge(s1, s3)\n",
    "show_clusters(otus, plot_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's cluster a fourth sequence, ``s4``. We find that this falls outside the range of ``OTU 1``, and also outside the range of ``OTU 2``. So, we'd create a new OTU, ``OTU 3``, containing ``s4``."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1 0.2 0.2\n",
      "OTU 2 0.5\n"
     ]
    }
   ],
   "source": [
    "for i, otu in enumerate(otus, start=1):\n",
    "    print(' '.join([\"OTU %d\" % i] + map(str,[s.fraction_same(s4) for s in otu])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['s1', 's3']\n",
      "OTU 2: ['s2']\n",
      "OTU 3: ['s4']\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAD9JJREFUeJzt3W9sXudZx/HvlcR/8ij+k7heu7bLuq4a1Qo0rdHEVApj\n",
       "aGoLbYVSpGCGpqmMDCa2rGBkbfhFJSpQkKW2K9tQ0CJRqYgIqZoStKWo5AVNKVTzSNickW6INiAq\n",
       "mjrEjuvEdpebF+fxmnSp/diJn3Me39+PFCnHPo91KS9+vnLd97lPpJSQJOVjXdkFSJKay+CXpMwY\n",
       "/JKUGYNfkjJj8EtSZgx+ScqMwS9JmTH4JSkzBr8kZcbgl6TMGPySlBmDX5IyY/BLUmYMfknKjMEv\n",
       "SZkx+CUpMwa/JGXG4JekzBj8kpQZg1+SMmPwS1JmDH5JyozBL0mZMfglKTMGvyRlxuCXpMwY/JKU\n",
       "GYNfkjJj8EtSZgx+ScqMwS9JmTH4JSkzG8ouYC2IiB6gr345kVKaLLMeSVqMHf8KRURHRAxujjjS\n",
       "Dif74Wg/HG2Hk5sjjkTEYES0l12nJL1dpJTKrqHlrIvY0Ql7boMYgq77eOu/TvPAAWAUzhyBdA52\n",
       "nk9pX3nVStLFDP5l2hjx0CZ45CDUBpa4dwy4G2amYeRsSo82oz5JWorBvwzrInb0wd4xqG1t8DMn\n",
       "gAGYmYAH7fwlVYHB36CI6NgIrx2G7tuX+dkx4E6YOgv9KaW51ahPkhrl4m7jtm+DaDT0f4niH/c8\n",
       "MADcCgFsX63iJKlRBn+DemF4CLoaufcp4E2KpF8wBF29MLwqxUnSMjjqaUBE9LTDyTegbWH3zm7g\n",
       "CWAKuBb4CvBRYBL4EPAk8GGKXwDrKHb7bIL5uWLc4z5/SaXxAa7G9PXA7AZoAzgOfBn4FnANxQLu\n",
       "m/Ubvwh8Brj6bT+gDeiGuddhC8XvB0kqhaOeFVgPzALjFJ38VuBGil8ELwCfLa80SVqSwd+YiUno\n",
       "mK9f3AQ8BjxM0dkPAv9D0ek/xsX/qAuDtHlgCtqBU80oWJLeicHfgJTSZA2OHbjga4PAc8ArFIu4\n",
       "n6bYtrkDeDfFnB/geuB5YD9Qg3Hn+5LKZvA36DTsHoUzAC8BhyjGPR1AJ3AV8CpwtP7nG/XPfZvi\n",
       "l8AonDldrAlLUqnc1dOgCx/gagM+BXyPYtH2DmAPxULvgpeB91OMeP4VH+CSVB0G/zJ4ZIOktcBR\n",
       "zzKcT2nfNIwMwMxYA/ePUYT+NIwY+pKqwo5/BRaOZd5WP5b5fi4+lnk/xUz/qMcyS6ogg3+F6i9Z\n",
       "2d4LwzNwSzfMQbFlswbj9YXcp53pS6oag/8KqL96cUv98pRbNiVVmcEvSZlxcVeSMmPwS1JmDH5J\n",
       "yozBL0mZMfglKTMGvyRlxuCXpMwY/JKUGYNfkjJj8EtSZgx+ScqMwS9JmTH4JSkzBr8kZcbgl6TM\n",
       "GPySlBmDX5IyY/BLUmYMfknKjMEvSZkx+CUpMwa/JGXG4JekzBj8kpQZg1+SMmPwS1JmDH5JyozB\n",
       "L0mZMfglKTMGvyRlxuCXpMwY/JKUmQ1lFyBJV1JE9AB99cuJlNJkmfVUkR2/pJYXER0RMbg54kg7\n",
       "nOyHo/1wtB1Obo44EhGDEdFedp1VESmlsmuQpBVbF7GjE/bcBjEEXffx1ihjHjgAjMKZI5DOwc7z\n",
       "Ke0rr9pqMPgltayNEQ9tgkcOQm1giXvHgLthZhpGzqb0aDPqqyqDX1JLWhexow/2jkFta4OfOQEM\n",
       "wMwEPJhz5++MX1LLiYiOTtjzzCKh/zfAzUAPcBWwnWIEdBBqnbAn55m/wS+pFW3fBnH7IjfcAfwj\n",
       "MAm8AtSA3wcGgFshKH4XZMngl9RyemF4CLoWrncD1wPdFF3+IeA9wLvq30/AeuDd9esh6OqF4eZV\n",
       "XC3O+CW1lIjoaYeTb0DbBuA48DHgReAaijn+m8CNwGHgXmAK+AXgGaCdYrfPJpifg/4c9/nb8Utq\n",
       "NX09MLuwZXM9MAuMUwT6VorQB/g54DTw30Ab8If1r7cB3TAHbGlW0VVi8EtqaTcBjwEPA1cDg8Cr\n",
       "b7vnWuCPgSebWll1GfySWs3EJHTMX/CFQeA5ikXc4NLD+3mKBd6Fv08VU59Tq1loVRn8klpKSmmy\n",
       "BscO1K9foljMnQU6gE6K8c9fU8z7ofiF8EfAA/Xr/UANxnOc74PBL6kFnYbdo3AGisD/AtBPsWvn\n",
       "deBPgWMUWzo3AR8BPgz8Wf3zo3DmdLEZKEvu6pHUciKiYyO8dhi6F9vLfyljwJ0wdbbY0TO3GvVV\n",
       "nR2/pJaTUpo9BzvvgpkTS9/+Iycozus5BztzDX0w+CW1qPMp7ZuGkQGYGWvg/jGKc3qmYSTnc3rA\n",
       "UY+kFlc/lvkvfwo2DUPcz8XHMu+nmOkf9VjmHzH4JbW8iPgD4IFe2DgDt9QfzmIK2mswXl/IfTrn\n",
       "8c6FDH5JLS0i1lPs6vzNlNIL9VcvLjyReyrXLZuL8Z27klrdvcAE8M9Q7POnOJRT78DFXUmtbhfw\n",
       "WHJ80TBHPZJaVkT8NPBN4H3O7xtnxy+plX0O+Kqhvzx2/JJaUkT0UyzqfiCldLLselqJHb+kVrWT\n",
       "Youmob9MdvySWk5EtAEvA/eklP6t5HJajh2/pFb0a8BLhv7KGPySWtEu4PGyi2hVBr+klhIRPwu8\n",
       "Cziw1L26NINfUqvZBTyRUvph2YW0Khd3JbWMiLgO+A7FA1sey7BCdvySWslngKcM/ctjxy+pJUTE\n",
       "Ror3pt+RUvp+2fW0Mjt+Sa3i48CLhv7lM/glVV5EBG7hvGIMfkmt4Bcp8urZsgtZCwx+Sa3g88CX\n",
       "PHP/ynBxV1KlRcT7Kd6u9d6U0kzZ9awFdvySqu6zwNcM/SvHjl9SZUVEN8UpnLemlP6r5HLWDDt+\n",
       "SVX2SeBZQ//KsuOXVEkRsR44DnwipfRPZdezltjxS6qqXwb+D3ih7ELWGoNfUlXtAh53C+eV56hH\n",
       "UuVExE8Cfw/ckFKaK7uetcaOX1IV7QK+auivDjt+SZUSEVcB3wd+IqX0Wtn1rEV2/JKq5reBrxv6\n",
       "q8eOX1JlREQb8J/AvSmlI2XXs1bZ8UuqkgeAHxj6q8vgl1QlnrnfBAa/pEqIiA8B1wD7y65lrTP4\n",
       "JVXFLuDPU0o/LLuQtc7FXUmli4jrgO8C70spnS67nrXOjl9SFfwu8JSh3xx2/JJKFRGdwCvAz6eU\n",
       "jpddTw7s+CWV7TeAMUO/eQx+SaWJiMAtnE1n8Esq00eANoqTONUkBr+kMu0CvuSZ+83l4q6kUkTE\n",
       "jcCLwHtTSm+UXU9O7PglleX3gL2GfvPZ8UtquojoAl4GbkspnSi5nOzY8UsqwyeBQ4Z+Oez4JTVV\n",
       "RKwD/h14MKV0uOx6cmTHL6nZ7gHOAM+XXUiuDH5JzfZ54HG3cJbHUY+kpomIW4BngRtSSrNl15Mr\n",
       "O35JzfQ54C8M/XLZ8Uu64iKiB+irX06klCYjYgvwH8DNKaX/La86bSi7AElrQ0R0ANt7YbgdPtgD\n",
       "swCT0LE54hhwDPg7Q798dvySLtu6iB2dsOc2iCHouo+3usp54ACwG9J3YOYc/Nb5lPaVV60MfkmX\n",
       "ZWPEQ5vgkYNQG1ji3jHgbpiZhpGzKT3ajPr04wx+SSu2LmJHH+wdg9rWBj9zAhiAmQl40M6/HAa/\n",
       "pBWJiI6N8Nph6L59mZ8dA+6EqbPQn1KaW4369M7czilppbZvg1gs9L8L3AX0c3HYDAC3QgDbV7E+\n",
       "vQODX9KK9MLwEHQtdk878OvA1y7xvSHo6oXhVSlOi3LUI2nZIqKnHU6+AW0Lu3d2A08AU8C1wFeA\n",
       "j9a/9wPgA8D5C37GPLAJ5ueKcc9kk0oX7uOXtDJ9PTC7oXhfLseBLwPfAq6hWMB9c4kf0AZ0w9zr\n",
       "sAUw+JvI4Jd02dZTPK01TvG4bqM7fFQOZ/ySVmJiEjrm6xc3AY8BDwNXA4PAq0v8gHlgqlgGOLVa\n",
       "RerSDH5Jy5ZSmqzBsQMXfG0QeA54hWK7zlKrtvuBGow7328+g1/SipyG3aPFC1V4CThEMe7pADop\n",
       "xj8A54CFjfqz9T8Ao3DmdLEmrCZzV4+kFbnwAa424FPA9ygWbe8A9lCE/o0L9wMJuAH4W3yAq0wG\n",
       "v6QV88iG1uSoR9KKnU9p3zSMDMDMWAP3j1GE/jSMGPrlseOXdNkWjmXeVj+W+X4uPpZ5P8VM/yik\n",
       "c7DT0C+XwS/pioiIduovYpmBW7rra7pT0F6D8fpC7tPO9Mtn8Eu64uqvXtxSvzzlls1qMfglKTMu\n",
       "7kpSZgx+ScqMwS9JmTH4JSkzBr8kZcbgl6TMGPySlBmDX5IyY/BLUmYMfknKjMEvSZkx+CUpMwa/\n",
       "JGXG4JekzBj8kpQZg1+SMrNh6Vv0TupvGeqrX074liFJrcCOf5kioiMiBjdHHGmHk/1wtB+OtsPJ\n",
       "zRFHImKw/u5RSaokX724DOsidnTCntsghqDrPt76L9M8cAAYhTNHIJ2DnedT2ldetZJ0aQZ/gzZG\n",
       "PLQJHjkItYEl7h0D7oaZaRg5m9KjzahPkhpl8DdgXcSOPtg7BrWtDX7mBDAAMxPwoJ2/pCpxxr+E\n",
       "iOjohD3PLBL6fwX8DNADvAcYBq4DDkKtE/Y485dUJQb/0rZvg7h9kRvOAo8DE8C/AP8AjAIDwK0Q\n",
       "wPZVr1KSGmTwL6EXhoega+F6N3A90A3cDBwCfge4g2Kh91rg48Dz9fuHoKu3+E+AJFWCM/5FRERP\n",
       "O5x8A9o2AMeBjwEvAtdQzPHfBG582+d+Ffgg8CcUu302wfwc9LvPX1IV2PEvrq8HZhe2bK4HZoFx\n",
       "ikDfyo+H/l7g28BQ/boN6IY5YMvqlytJSzP4l+Em4DHgYeBqYBB49YLvfx34IvBNTHlJ1eWoZxEL\n",
       "o55paGt72/fOAJ+mmOs/CRwEPgF8g2KHzwJHPZKqxo5/ESmlyRocO1C/foliMXcW6AA6KcY/hygW\n",
       "dJ/m4tAH2A/UYNzQl1QVBv8STsPu0aLBZxb4AtAPvBt4nWIB9xGKG+6h2P7TBfxK/fOjcOZ0sRlI\n",
       "kirBUc8SIqJjI7x2GLoX28t/KWPAnTB1thjzzK1GfZK0XHb8S0gpzZ6DnXfBzIllfO4ExXk952Cn\n",
       "oS+pSgz+BpxPad80jAzAzFgD949RnNMzDSOe0yOpahz1LMPCsczb6scy38/FxzLvp5jpH/VYZkkV\n",
       "ZvAvU/3Ate29MDwDt9QfzmIK2mswXl/IfdrxjqSqMvgvQ/3ViwvPap1yy6akVmDwS1JmXNyVpMwY\n",
       "/JKUGYNfkjJj8EtSZgx+ScqMwS9JmTH4JSkzBr8kZcbgl6TMGPySlBmDX5IyY/BLUmYMfknKjMEv\n",
       "SZkx+CUpMwa/JGXG4JekzBj8kpQZg1+SMmPwS1JmDH5JyozBL0mZMfglKTMGvyRlxuCXpMwY/JKU\n",
       "GYNfkjJj8EtSZgx+ScqMwS9JmTH4JSkzBr8kZcbgl6TMGPySlBmDX5IyY/BLUmYMfknKjMEvSZkx\n",
       "+CUpMwa/JGXG4JekzBj8kpSZ/wdGuXtZxuoDOQAAAABJRU5ErkJggg==\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x106629d10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "G=nx.Graph(id=\"OTU 3\")\n",
    "G.add_node(s4)\n",
    "otus.append(G)\n",
    "show_clusters(otus, plot_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, let's cluster our fifth sequence, ``s5``. We find that this falls outside the range of ``OTU 1``, but inside the range of both ``OTU 2`` and ``OTU 3``. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1 0.4 0.4\n",
      "OTU 2 0.7\n",
      "OTU 3 0.8\n"
     ]
    }
   ],
   "source": [
    "for i, otu in enumerate(otus, start=1):\n",
    "    print(' '.join([\"OTU %d\" % i] + map(str,[s.fraction_same(s5) for s in otu])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Algorithmically, we now have a choice to make. How do we decided which OTU a sequence should belong to if it is within the similarity range of several OTUs? \n",
    "\n",
    "A few choices would be:\n",
    " 1. add it to the cluster where it first matches all sequences, \n",
    " 2. add it to the cluster where it has the smallest average distance to the cluster members, \n",
    " 3. or add it to the cluster with the most members. \n",
    "\n",
    "There are many other options as well. Let's choose the option 1 here, as it requires performing the fewest number of alignments, so should be fastest (we'll explore that in more detail soon). Our mapping of OTUs to sequences would look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['s1', 's3']\n",
      "OTU 2: ['s5', 's2']\n",
      "OTU 3: ['s4']\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAEcpJREFUeJzt3X+s3XV9x/Hnu+29t/fa9raUCp1YtTLHwKWFLhpg1alj\n",
       "ZXF2pgt0V6PZGq1mi0tculVnF8nG1G5VXJy6ksFkZEjNYBM2R4zRRQibhIttZlFxk9E4iC2t7b2X\n",
       "295e6Gd/fL6He1vuj3OLt+d77uf5SG56z/eeL3wuKa/v5/v+fj7vEyklJEnlmNfqAUiSzi2DX5IK\n",
       "Y/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAG\n",
       "vyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BL\n",
       "UmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwixo9QCksxURvcDy6uXhlNKxVo5HahfO+NVWIqIrIvqW\n",
       "RezthEMrYN8K2NcJh5ZF7I2IvojobPU4pTqLlFKrxyA1ZV7E5oVw8+UQ22Dx2xi7ZR0F7gV2weBe\n",
       "SCdg66mU9rRutFJ9GfxqC90RH1wEN94HPeumeW8/cC0MD8GO4ynddC7GJ7UTg1+1Ny9i83K4tR96\n",
       "VjV5zgFgHQwfhi3O/KXTGfyqtYjo6oaDD8CSK2Z4bj+wHgaOw4qU0snZGJ/Ujny4q7rbtBZiqtD/\n",
       "AjAfWDzu65vAOmANBLBptgcptRODX7W2FLZvy1k+pauBwXFfb6iOb4PFS2H7LA5RajsGv2orInqH\n",
       "4dKN447tBC4ClgCXAF+vjk9WsNwIDMNl1Zp/SRj8qrflvTDSWLL5feCzwMPAAPBV4JXkWs63gRXA\n",
       "zwE3As9V53QAS+AkcN65G7ZUb+7cVduYD4wA+8nbdRsrfKI69grgO8Bm8l/sD7VgjFI7cFWPaisi\n",
       "ejvh0BB0dFTHvgh8jhz0G4BPASvPOG8P8JfkO4NRYBFwMleFHgN+CPxP9ecPU0oDs/6LSDVjqUe1\n",
       "lVI61gOP3jvuWB9wP/AEeaY/2VPbxnTmHqAbvgvsIl8vVgLvBm4HnoqIQxHxrYj4YkT8eURsiYhf\n",
       "johVETF/Nn4vqdWc8avWIqLvStj9ICx+DPgReQVPAO8nB/xm4HLgAuB7wHXA9cCfAFfC4H/C1pTS\n",
       "nRP8swN4KfBqYPW4r8br5eRrzGl3CdX3j6eUhmbr964rG+PNDQa/am38Bq4O4D3k6XsH+QKwm1zu\n",
       "uR0YIof/u8ihv5cXt4ErIrqBV/HCC8Lq6vggL7wgNL5/KqV06mx/7zqJiC5g01LYPgyX9uZHLRyD\n",
       "rh549GhebHWXm+Tah8Gv2qtjy4aImAdcyAsvCI3XvcDjTH63cPynPabZYGO8ucngV1totyZtEfES\n",
       "8l3B+ItC4/tXAkeY+KLwQ+DHqQb/Y7bbf3M1z+BX22jMPtdWs8+NnD77vIc8+9xX89lndbfwMiYu\n",
       "Ia0GXsLYxeDMi8PjKaWR2R5jHe+y9NNj8KutVB+y0qg3X1ZtzmIAOntgf1Vvvrud680RsYSxZwtn\n",
       "3jGsAg4y+d3C0y/2bqGZxni3AZ8BfkDeRf0O4GO8+OcqOjcMfrWtaoVJY0fukRJWmFRLTC9i4hLS\n",
       "avJz78nuFp5oJozHr6Sa7D1/A/wC8HryVWgjeTXVdqZeSaV6MPilOSQiljJ5Ceki4Ckmv1v4SUop\n",
       "LYvYewusabQ03Ume3Q8AP0PeQPfmM/69NwHfIJfb7gLeA3t/ktLls/V76sUx+KVCREQH8HImLiG9\n",
       "mrwt4vFOWPMMxAJyf6RrgIfIS5gOAM9WJ433duBScrmn2i09ejKXe+b8XVg7slePVIiU0ihjs/yv\n",
       "jf9ZtZltGbC+F+5cAAth8v5I490KPFL9CWON8Z7OZTiDv4Zs2SCJlB0B/os8qQfgYuDTwA3kzXF9\n",
       "5FpRwz8Dfwz8G7Y/bSeWeiQ9b6LGeA2DwPvIZYK/B+4jNz36CvCL495nqaf+nPFLet6ZjfEeI7c1\n",
       "HQG6yPWf+dWxdwJ3c3roQ37A2wP7Df36MvglneYo7NyVJ/iMAB8mf8jNSuBp8gPcG8lv+DXGPuf4\n",
       "rdX5u2Cw2k+hmrLUI+k0zWzgmkw/buBqB874JZ0mpTRyArZugOEDMzjvALlfz4m8ecvQrzGDX9IL\n",
       "nEppzxDsWAfD/U28v5/cp2cIdtinp/4s9Uia1FxpjKfTGfySplRCY7zSGPySmlZiY7y5yOCXpML4\n",
       "cFeSCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4\n",
       "JakwBr8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+S\n",
       "CmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4Jakw\n",
       "Br8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPw\n",
       "S1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8k\n",
       "Fcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1Jh\n",
       "DH5JKozBL0mFMfglqTAGvyQVZkGrB9DOIqIXWF69PJxSOtbK8UhSM5zxz1BEdEVE37KIvZ1waAXs\n",
       "WwH7OuHQsoi9EdEXEZ2tHqckTSZSSq0eQ9uYF7F5Idx8OcQ2WPw2xm6ZRoF7gV0wuBfSCdh6KqU9\n",
       "rRutJE3M4G9Sd8QHF8GN90HPumne2w9cC8NDsON4Sjedi/FJUrMM/ibMi9i8HG7th55VTZ5zAFgH\n",
       "w4dhizN/SXVi8E8jIrq64eADsOSKGZ7bD6yHgeOwIqV0cjbGJ0kz5cPd6W1aCzFV6N8JXAL0AucD\n",
       "m4AngXXAGojqkCTVgsE/jaWwfRssnuo9VwPfBI4BTwA9wB9UP9sGi5fC9tkdpSQ1z+CfQkT0DsOl\n",
       "G8cd2wlcBCwhz/K/DrwceGn18wTMB1ZWrzcCw3BZteZfklrO4J/a8l4YaSzZ/D7wWeBhYAD4KvDK\n",
       "6mcPAEvJF4QD5AsEQEc+dhI47xyNWZKmZPDPwHxgBNhPXre/Clhd/eyXgKPAj8hh/4etGKAkNcFV\n",
       "PVOIiN5OODQEHR3VsS8CnyOH/wbgU4yVdRq+BVwL/IR8gVgEoyfzyh5bOkhqOWf8U0gpHeuBR+8d\n",
       "d6wPuJ/8EDeY+KntKPkBL8A9+fv9hr6kujD4p3EUdu6CQYDHyA9zR4AuYCG5/HMHua4P+YLwEeA3\n",
       "q9e7YPDoWMlfklrOUs80xm/g6gDeA3yXXMe/GrgZ+GvgNnJpZwWwGbiBXA5yA5d07tk5d2oGfxPO\n",
       "tmXDFXDqcF4EtCGldHQWhygVLyK6gE1LYfswXNqbb845Bl098Gh1532XkzBLPU05ldKeIdixDob7\n",
       "m3h/P7lPzzPwIeAhoD8ipuvtJukszYvY3A0Hr4Ldt8KaZ6DjICw6CIuGoOMWWHMl7O6GQ/MiNrd6\n",
       "vK3mjH8GGm2Z11ZtmTdyelvme8g1/X1ntGWOiOvIi4FuAD6X/I8u/dTYOXfmDP4Zqj5kpXE7eVm1\n",
       "OYsB6OyB/dXt5N1n3k5GxM8CXyI/I35vSmngnA9emmPsnHt2DP4XoXqA1NiRe2S6B0gRsRD4NPBm\n",
       "4LqU0r5ZHqI0Z820c+5bgG8AzwLfpuyFF9b4X4SU0rGU0uPV17SrBlJKJ1JK7yeXfL4WEe+NiJj1\n",
       "gUpz07Sdcxv+gRz4jf/ZSu+ca/C3QErpDmA98PvA7RGxqMVDktrOmZ1zJ2qgCLlr7p8Cf0FuothQ\n",
       "cudcSz0tFBE95G0AV5JLP99p8ZCkttBop/IMdCwgN1C8hryE7kJyHf9Zci+t3wNeA/xG9fpZ8oy3\n",
       "5HYqzvhbKKU0nFLaAnwC+EZE/HaLhyS1i9M6507WQPFh4D+AD0zwDyi5c67BXwMppduANwHbI+Lv\n",
       "qjsBSU26mLxq4gbgAnJPrSeB362Ojw86axyWemqlqvV/HlhLLv18r8VDkmppos65DYPA+6o/v8LY\n",
       "hyQ9BzxNvjD8I/A6LPWoBlJKQ8C7gb8C7o+Id7Z4SFItndk5d6IGiucDTwH7qq+vVO99hBz6JXfO\n",
       "NfhrJmV/C/wK8NGI2B0R3a0el1Q34zvnjgAfJjdJXEme2X+cPNtvfJ1PXr95Abm+X3LnXEs9NRYR\n",
       "S8gNQC8hl35+0OIhSbUx0w1c4/XjBi7VVNXWoQ/YDTwYEde3eEhSbaSURk7A1g0wfGD6tz/vALlf\n",
       "zwnYWmLog8Ffe1Xp5/PkT3r8WER8tmo/KxXvbDvnDsGOUvv0gMHfNlJKj5B3ml9Inv2vnuYUqQjH\n",
       "U7rpMGxZDwNXweDd5E1aDaPAXcCVMLgeBg7DlpI7c4I1/rZT9fb5ALADeF9K6Z9aPCSpFs62c26J\n",
       "DP42FRGvA/YAXwb+yL/M0piZds4tjcHfxiLiPOAL5BVq16eUnmjtiCS1A2v8bSyldITce+pLwEMR\n",
       "8bYWD0lSG3DGP0dExFXAndXXR1JKoy0ekqSacsY/R6SUHgSuAF4L/HtEXNTiIUmqKYN/DkkpPQ38\n",
       "OvAvwMMRcW2LhySphiz1zFER8QbgDuA24KMppWenOUVSIQz+OSwiXkr+uNEO4B0ppSdbPCRJNWCp\n",
       "Zw5LKR0EriV3rH04It7S4iFJqgFn/IWoQv92crfPP0spPdfiIUlqEYO/IBGxklz3fw54Z0rpxy0e\n",
       "kqQWsNRTkJTSU+QPeHkQ6I+IN7Z4SJJawBl/oSLiV8krfj4DfCKldKrFQ5J0jhj8BYuIl5F3+g4B\n",
       "76r2AUia4yz1FCyl9H/Am8ifRf1IRFzd4iFJOgec8QuAiHgrcAvwSeCT05V+qra3y6uXh217K7UP\n",
       "Z/wCIKX0r8DrgE3Al6uWz6eJiK6I6FsWsbcTDq2AfStgXyccWhaxNyL6qg/DkFRjBr+el1I6ALwR\n",
       "eIxc+nl942fzIjZ3w8GrYPetsOYZ6DgIiw7CoiHouAXWXAm7u+HQvIjNLfslJE3LUo8mFBFvJ2/2\n",
       "+lgXxGK48T7oWTfNef3AtdWHWZf+uaZSXRn8mlREvAr42nJ4xSMwf1WT5x0A1sHwYdhyKqU9szhE\n",
       "SWfBUo+m8mQ3nP/VKUL/O8AGYAVjf5lWAfdBz0K42Zq/VD8Gv6ayaS3EFVO8oRP4LfJyoPHWAWsg\n",
       "yA+LJdWIwa9JLYXt22Bx4/VO4CJgCXAJueXna4DfAS6d4PxtsHgpbD8XY5XUPGv8mlBE9HbCoWeg\n",
       "YwHwfeAa4CHgQnId/1lgdfX+/yZfBMYv/h8FFsHoSVjhOn+pPpzxazLLe2FkQfViPjAC7CcH+irG\n",
       "Qn8yHcASOAm8YE+ApNYx+NWUi4FPAzcAFwB9wFOtHJCks2bwazKHj0HX6LgDfcD9wBPkp7bTFe9H\n",
       "gYH8/PfI7AxR0tkw+DWhlNKxHnj03ur1Y+SHuSNAF7CQXP4BOEGu51D9fKT6/h6gB/Zb35fqxeDX\n",
       "pI7Czl0wCDnMP0xer78SeBr4OPC/QA/wWvJdQDfw89X5u2DwaF4MJKlGXNWjSUVEVzccfACWTLWW\n",
       "fyL9wHoYOJ5X9Jyc9gRJ54wzfk0qpTRyArZugOEDMzjvALlfzwnYauhL9WPwa0qnUtozBDvWwXB/\n",
       "E+/vJ/fpGYId9umR6slSj5oyL2LzQrh5LcQ2WLwRaKzxHyU/yN0Fg/sgnYCthr5UXwa/mlY1XNu0\n",
       "FLYPw2XV5iwGoLMH9lcPcu+2vCPVm8Gvs1J99GJjR+4Rl2xK7cPgl6TC+HBXkgpj8EtSYQx+SSqM\n",
       "wS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8\n",
       "klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9J\n",
       "hTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQY\n",
       "g1+SCmPwS1Jh/h+bQa8ZBfSKQwAAAABJRU5ErkJggg==\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10062b210>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "otus[1].add_edge(s2, s5)\n",
    "show_clusters(otus, plot_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, let's cluster our last sequence, ``s6``. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1 0.8 0.6\n",
      "OTU 2 0.4 0.3\n",
      "OTU 3 0.2\n"
     ]
    }
   ],
   "source": [
    "for i, otu in enumerate(otus, start=1):\n",
    "    print(' '.join([\"OTU %d\" % i] + map(str,[s.fraction_same(s6) for s in otu])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case, it falls within the similarity range of ``s1``, but outside of the similarity range of ``s3``. So, because our algortihm requires that a sequence be within the simiarity range of all sequences in an OTU, ``s6`` cannot be a member of ``OTU 1``, so instead it's assigned to a new OTU, ``OTU 4``. Our final mapping of OTUs to sequences would look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['s1', 's3']\n",
      "OTU 2: ['s5', 's2']\n",
      "OTU 3: ['s4']\n",
      "OTU 4: ['s6']\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAEvhJREFUeJzt3XuQ3WV9x/H3NyS7yZorIaAVqabIIKABVukgTccpteBQ\n",
       "UeNIGkWnTUfUVrFo2gim1gtCo6lorZ2SFkYdrDCUaMmMxXtH0dZLNGlJVKi31EslhJILuS3m2z+e\n",
       "33Y3muyeTTZ7ztnn/ZrZIee3v/Obh8nmc579/p7f94nMRJJUjyntHoAkaWIZ/JJUGYNfkipj8EtS\n",
       "ZQx+SaqMwS9JlTH4JakyBr8kVcbgl6TKGPySVBmDX5IqY/BLUmUMfkmqjMEvSZUx+CWpMga/JFXG\n",
       "4Jekyhj8klQZg1+SKmPwS1JlDH5JqozBL0mVMfglqTIGvyRVxuCXpMoY/JJUGYNfkipj8EtSZQx+\n",
       "SaqMwS9JlTH4JakyBr8kVWZquwegySEi5gDzm5fbM3NHO8cj6cic8euoRURvRCybF7GxB7YtgE0L\n",
       "YFMPbJsXsTEilkVET7vHKelQkZntHoO60JSIpdNh7XkQK2DW8xn69XEAWA+sgV0bIffBVQcz72jf\n",
       "aCUNZ/BrzGZEXDMTrr8H+vpHOXcDcCns2Q2r9mbeNBHjkzQyg19jMiVi6Xy4dQP0ndbie7YC/bBn\n",
       "Oyx35i+1n8GvlkVE7wx48F6Yff4Y37sBWAw798KCzDxwPMYnqTXe3NVYLDkXYqTQvx04E5gDnAQs\n",
       "AX4C9AOLIJpDktrI4FfL5sLKFTBrpHMuAr4A7AB+CPQBb2i+twJmzYWVx3eUkkZj8KslETFnD5x1\n",
       "+bBjq4FTgdmUWf7ngCcBJzffT+AE4AnN68uBPXB2s+ZfUpsY/GrV/Dmwf3DJ5neADwBfB3YCnwKe\n",
       "3HzvXmAu5QNhK+UDAmBaOXYAOHGCxizpMAx+HZUTgP3AZsq6/dOAhc33fgN4BPgRJez/tB0DlHRE\n",
       "rupRSyJiTg9s2w3TpjXHPgr8LSX8LwHew1BZZ9BXgEuB/6V8QMyEgQNlZY8tHaQ2ccavlmTmjj7Y\n",
       "sn7YsWXAFyk3cYPD37UdoNzgBbi7/HmzoS+1l8Gvlj0Cq98NewHup9zM3Q/0AtMp5Z9/pNT1oXwg\n",
       "vBl4cfN6Dex6ZKjkL6lNDH61JCKeA1y9CXq/QQn8a4EFlPLOQ8CNwBbKks6ZwHOAC4F3UR7g2lQW\n",
       "+qyb6LFLOpQ1fo0oIp4FvBP4NeAvAh6bD7fYskHqXs74dVgRcXZErAM+RpmlPy0zbzuYeftuWNUP\n",
       "eza0cJ0NlNDfDasMfakzGPw6REQsjIgPA58HvgQ8NTP/bnh/nb2ZN22H5Yth57Nh1zrgsWHXGADu\n",
       "Ai6EXYth53ZYbmdOqXNY6hEAEfFEYBXwEuD9wE2ZuXOU9/QAS+bCyj1wdvNwFjuhpw82Nzdy19mU\n",
       "TeosBn/lIuIkykrMPwRuAVZn5kNHcZ05DD2R+7BLNqXO5Z67lYqI2cA1wOuAO4GnZ+aPj/Z6TdAb\n",
       "9lIXsMZfmYiYERFvBB6grNS5IDNfcyyhL6m7OOOvRERMo5RzVgFfBS7OzPvaOyrp8JrS4fzm5XZL\n",
       "h+PLGf8kFxEnRMSVwLcpm6C8KDOXGPrqNBHRGxHL5kVs7IFtC2DTAtjUA9vmRWyMiGXNggIdI2/u\n",
       "TlIREcALgOspnZPfnJmfb++opMObErF0Oqw9D2IFzHo+Q+WIAWA9peXHRsh9cJXPhBwbg3+SaQL/\n",
       "YuAGoIfSLucT6V+0OtSMiGtmwvX3QF//KOduAC5tHgj02ZCjZ/BPIhFxIaW9wqnAnwN3ZubB9o5K\n",
       "OrIpEUvnw622AJlYBv8kEBHPoAT+IuBtwIcy87GR3yW1V0T0zoAH74XZ549w3veAqyl7OfcCy4Er\n",
       "gMWwc2/Z28EHBMfIm7tdLCLOiIiPUnY+/AxwRmbeYuirSyw5F2Kk0D8APBf4beBnwI+BK4F+YFHZ\n",
       "BmLJcR/lJGTwd6GIeFJE/D3wZeA+4PTMfF9m7mvz0KSWzYWVK2DW4OvVlBrlbOBMyn4PH2yO/Qkw\n",
       "g3LT6unN+Stg1tzD7/+jUVjq6SIRcTKlDf4rgJuBNZn5cHtHJY3d4Faej8K0qcB3KDP7rwKPp9Tx\n",
       "H6PULwco+z18DTiH0kjqHNzK81g44+8CETE3It4BfIuy0dXZmXmdoa8uNn8O7B9csnkCZXOfzZRA\n",
       "Pw1YCPw3cDvweuCnwGWUNcoDwDSgaQx4IhoTg7+DRcTjIuJNlPYKTwT6M/PqzPyfNg9NGlenA+8F\n",
       "3gqcQtnP+aeU/ZoXA5dQ1vWvALZTnkbU0TP4O1DzBONrKYF/PrA4M5dn5g/aOzJp3GzfAb0Dww4s\n",
       "A75I2as5KMX7Z/zCm4YXpgcoLcABf/MdI4O/g0TE1Ij4fUrJ83nAZZl5RWY6wdGkkpk7+mDL+ub1\n",
       "/ZSbufspSzanU8o/VwL/DnwW+Dnlt4IFwNOAu4E+2Gx9f+wM/mMQEXOaHasWNk2ljvY6UyLiJcB/\n",
       "UpYpX5mZl2XmN8dtsFKHeQRWr4FdUAL/WkqoP4FyM/dG4AzgNuDVlEL+ekrgT6W0cGg2+9EYuapn\n",
       "jCKil6Fdp86aU35m2QG9fbCl+UG8q5WHSpr2CpdSFi8kcB3wKdsrqAatPsB1OBvwAa5j4Yx/DKZE\n",
       "LJ0BDz4bbr4VFj0K0x6EmQ/CzN0w7RZYdCHcPAO2TYlYOtK1ImIx5WHE91CC/5mZ+UlDX7XIzP37\n",
       "4KpLYM/WMbxvK6Vfzz64ytA/Os74WzRejaQiop/SMfNMyiKG2zLz58dl0FIXsEnbxDP4WzAejaQi\n",
       "4mnA24GLKMH/D85WpGKwLfO5TVvmyzm0LfPdlJr+JtsyjwuDfxSt1CHvA94IfIOyxniwHWZTh9y1\n",
       "Fz5OWaXzbuBvMnPP8R631G2aTVYG75+d3TycxU7o6YPNzf2zdU6Yjp3BP4qIWHYh3PzlYT1FftH9\n",
       "wJco+8S9kKHgB7gA8mvwMWC5y86k1jSr5AafyH3Yfzvjy5u7o2ilkdQZwB8AZx3m/Ssh5sJCf3Cl\n",
       "1mXmjsz8fvPlv51x5ox/BK02klrYnP9flA+B4TN+G0lJ6jTO+EfWUiOpkdhISlKnMfjH4EiNpCSp\n",
       "mxj8I2upkdRIbCQlqdMY/CNotZEUwD6atWfN9/c3f7aRlKROY/CPopVGUj+g9A0/h/JbwAxK90Cw\n",
       "kZSkzuOqnlHYSErSZOOMfxQ2kpI02Rj8LTiYecduWNUPeza0cP4GSp+e3bDKniKSOo2lnjGwkZSk\n",
       "ycDgHyMbSUnqdgb/MbCRlKRuZPBLUmW8uStJlTH4JakyBr8kVcbgl6TKGPySVBmDX5IqY/BLUmUM\n",
       "fkmqjMEvSZUx+CWpMga/JFXG4Jekyhj8klQZg1+SKmPwS1JlDH5JqozBL0mVMfglqTIGvyRVxuCX\n",
       "pMoY/JJUGYNfkipj8EtSZQx+SaqMwS9JlTH4JakyBr8kVcbgl6TKGPySVBmDX5IqY/BLUmUMfkmq\n",
       "jMEvSZUx+CWpMga/JFXG4Jekyhj8klQZg1+SKmPwS1JlDH5JqozBL0mVMfglqTIGvyRVxuCXpMoY\n",
       "/JJUGYNfkipj8EtSZQx+SaqMwS9JlTH4JakyBr8kVcbgl6TKGPySVBmDX5IqY/BLUmUMfkmqjMEv\n",
       "SZUx+CWpMga/JFXG4Jekyhj8klQZg1+SKmPwS1JlDH5JqozBL0mVMfglqTIGvyRVxuCXpMoY/JJU\n",
       "mantHoAkjZeImAPMb15uz8wd7RxPp3LGL6mrRURvRCybF7GxB7YtgE0LYFMPbJsXsTEilkVET7vH\n",
       "2UkiM9s9Bkk6KlMilk6HtedBrIBZz2eojDEArAfWwK6NkPvgqoOZd7RvtJ3D4JfUlWZEXDMTrr8H\n",
       "+vpHOXcDcCns2Q2r9mbeNBHj62QGv6SuMyVi6Xy4dQP0ndbie7YC/bBnOyyvfeZvjV9SV4mI3umw\n",
       "9pNHCP0PAicAs4Z9fQE4DbgH+qbD2tpr/ga/pG6z5FyI80c44SJg17Cv32yO9wOLIIAlx3eInc3g\n",
       "l9RV5sLKFWUiz2rgVGA2cCbwueackQrYK2DWXFh5fEfZ2azxS+oaETGnB7Y9CtO+CzwX+CrweEoN\n",
       "/zHgi8AfAzOAE4GXA9dSyj9QVvvMhIEDsKDWdf4+wCWpm8yfA/unwrQTgP3AZsoTW4P1/miO/Spw\n",
       "H7CUEnRvar4/DZgNBx4qnwtVBr+lHkld6XTgvcBbgVOAZcBPgadQQh/gHOAtwD+1YXydzOCX1E22\n",
       "74DegebFMkpp54eUmf6RCvfDC9oDwE7oAR4+bqPscAa/pK6RmTv6YMt64H7Kzdz9QC8wnVLHvwf4\n",
       "WXP+t4HrgRcOu8bdQB9srrW+Dwa/pC4SEfMfgR+thtxPuWm7AHgC8BBwA/BZYBEwE7gMeDFw3bBr\n",
       "rIFdj5QFQdVyVY+kjhcRjwNeD7wBuHMGvOxemDXSWv7D2QAshp17y4qeA+M+0C7hjF9Sx4qIaRHx\n",
       "KkplZxFwYWa+Zh+88hLYs3UM19pK6dezD66qOfTB4G+7iJgTEQubrzntHo/UCaJ4MWVF5hXACzJz\n",
       "aWY+AHAw847dsKof9mxo4XobKH16dsOq2vv0gKWetoiIXmDJXFi5B86aU+5PsQN6+2BLU3+8q/ZZ\n",
       "ieoUEc+h/BuYRll+/+k8QlANtmU+t2nLfDmHtmW+m1LT32Rb5kMY/BPM/uHS4UXEIuBGSveFNwN3\n",
       "ZObBFt7Xw9BE6uzZcADKks0+2NxMpNY5kRpi8E8g+4dLvywingy8g9KB4Xpg7dGGdFMuPbF5+XDN\n",
       "SzZHYvBPEPuHS4eKiAWUmf3LgfcDf5WZu9o7qjp4c3cCjNY/fLiLKX8pB7F/uCaniHhcRKwCvkV5\n",
       "5uqszHyroT9xDP6JMWr/cICPULoLxrBj9g/XZNEszXwN8ABwFvDrmfm6zPzZKG/VODP4J0Ar/cN3\n",
       "AG8H3sUv9xK3f7i6WbM08wpgC/Ai4Hcz86WZ+d02D61a1viPs1b6hy+k9A8/A3hB8/oxhj6V7R+u\n",
       "bhURv0WZ7wTwpsz8TJuHJJzxT4TB/uEM7x8+QKnhLwS+Dvwb8LojXGCwfzhDqxWkjhYR50XEJ4G1\n",
       "wBrgAkO/cxj8E+hw/cN/AvxRc3z4X4a/h6kbNU+gfwT4BPDPlBu3La3H18Sx1HOcDZZ6dsO0acOO\n",
       "7wJe1fz3E8DJzfGfU7oMnkLZPOIiLPWo80XEycAq4KXAXwPvyczd7R2VjsStF4+zzNwxL2LLelh0\n",
       "DvAjSpgP9g/vpewaNGgrcAHwDeCk5thg//D9hr46TETMonTMvBq4jTLDf7C9o9JoLPVMgEdg9RrY\n",
       "dbj+4TdSZvuDXydR7oKdQqntg/3D1XkioiciXkvpmvlU4FmZ+XpDvztY6pkAEdE7Ax68F2bbP1zd\n",
       "LCKmULplvpOyHv/azPxme0elsXLGPwEyc/8+uMr+4epmEfFcyiK0NwKvzMxLDf3uZPBPEPuHq1tF\n",
       "RH9EfBr4AKU6eUFmfm6Ut6mDGfwTaG/mTdth+WLY+WzYtY7yoNagAeAu4ELYtRh2boflduZUu0TE\n",
       "6RFxO6Vb+F3A2Zl555F646t7WONvA/uHq5NFxCnAW4ClwE3AezPz0faOSuPJ4G8z+4drPDQ/R/Ob\n",
       "l9uP5ucoImZT6vevBT4MvDMzHxq/UapTWOpps8zckZnfb74MfbUsInojYtm8iI09sG0BbFoAm3pg\n",
       "27yIjRGxrJV23s11rqYszXwK0J+Z1xj6k5czfqkLjccWns3SzN+j7Hr1LcrSzP+YqP8HtY/BL3WZ\n",
       "Y93CMyIC+B3gLyn3l1Zm5r8e10Groxj8Uhc51i08E75HCfxTKQ+Sf8xVOvUx+KUu0coT4B+ibF77\n",
       "AGWzn5cCNwAbgcUwsBe2AW8Dbs3Mx45wGU1y3tyVuseoW3juBd4HbAe+AnyW0gy/H3h62cr52sxc\n",
       "a+jXzeCXusTwLTzh8Nt4vprS/XUq8CvAy4AvNef/GfTOhWsmdNDqSJZ6pC4wfAvPqcB3OPI2nsO9\n",
       "kLKr+Q24r4OGOOOXusP/b+EJR97Gc7hbKfs6rGheu4WnBhn8Uhc63Daewzf0+ThwHfAvmPL6ZZZ6\n",
       "pC5wpC08YWgbz6mUPgv3AK+gbOn5zGHnWerRIGf8UhfIzB19sGV98/p+ys3c/Qxt43lCc+xlwDoO\n",
       "DX0Y2sLT0JfBL3WJwS08oQT+L27jeQOl98Iu4HmU5T+zgMua97uFpwZZ6pG6hFt4arw445e6hFt4\n",
       "arwY/FIXcQtPjQdLPVIXGmzLfG7TlvlyDm3LfDelpr9phLbMqpfBL3Upt/DU0TL4pUnALTw1Fga/\n",
       "JFXGm7uSVBmDX5IqY/BLUmUMfkmqjMEvSZUx+CWpMga/JFXG4Jekyhj8klQZg1+SKmPwS1JlDH5J\n",
       "qozBL0mVMfglqTIGvyRVxuCXpMoY/JJUGYNfkipj8EtSZQx+SaqMwS9JlTH4JakyBr8kVcbgl6TK\n",
       "GPySVBmDX5IqY/BLUmUMfkmqjMEvSZUx+CWpMga/JFXG4Jekyhj8klQZg1+SKmPwS1JlDH5JqozB\n",
       "L0mVMfglqTIGvyRVxuCXpMoY/JJUGYNfkirzfzjDrjJ56A7cAAAAAElFTkSuQmCC\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1069c6f90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "G=nx.Graph(id=\"OTU 4\")\n",
    "G.add_node(s6)\n",
    "otus.append(G)\n",
    "show_clusters(otus, plot_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from numpy import mean\n",
    "\n",
    "def further_neighbor_cluster_match(seq, cluster, similarity_threshold, aligner):\n",
    "    alignment_results = []\n",
    "    for node in cluster.nodes_iter():\n",
    "        aln = aligner(seq, node)\n",
    "        percent_similarity = aln[0].fraction_same(aln[1])\n",
    "        alignment_results.append((node, percent_similarity))\n",
    "        if percent_similarity < similarity_threshold:\n",
    "            return False, alignment_results\n",
    "    return True, alignment_results\n",
    "            \n",
    "\n",
    "def furthest_neighbor_cluster(seqs, similarity_threshold, aligner=local_pairwise_align_ssw):\n",
    "    clusters = []\n",
    "    num_alignments = 0\n",
    "    for query_seq in seqs:\n",
    "        clustered = False\n",
    "        for cluster in clusters:\n",
    "            clustered, alignment_results = further_neighbor_cluster_match(\n",
    "                query_seq, cluster, similarity_threshold, aligner)\n",
    "            num_alignments += len(alignment_results)\n",
    "            if clustered:\n",
    "                break\n",
    "        if clustered:\n",
    "            for n, s in alignment_results:\n",
    "                cluster.add_edge(query_seq, n, percent_similarity=s)\n",
    "        else:\n",
    "            # create a new cluster containing only this node\n",
    "            new_cluster = nx.Graph(id=\"OTU %d\" % (len(clusters) + 1))\n",
    "            new_cluster.add_node(query_seq)\n",
    "            clusters.append(new_cluster)\n",
    "    return clusters, num_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "OTU 1: ['s1', 's3']\n",
      "OTU 2: ['s5', 's2']\n",
      "OTU 3: ['s4']\n",
      "OTU 4: ['s6']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/caporaso/.virtualenvs/iab/lib/python2.7/site-packages/skbio/alignment/_pairwise.py:540: EfficiencyWarning: You're using skbio's python implementation of Needleman-Wunsch alignment. This is known to be very slow (e.g., thousands of times slower than a native C implementation). We'll be adding a faster version soon (see https://github.com/biocore/scikit-bio/issues/254 to track progress on this).\n",
      "  \"to track progress on this).\", EfficiencyWarning)\n"
     ]
    },
    {
     "data": {
      "image/png": [
       "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\n",
       "AAALEgAACxIB0t1+/AAAFExJREFUeJzt3X+QnVV9x/H3N2Q37LrJJiwxYC110KKCTIA4bRHR1tZC\n",
       "i1CNtTH2hxXT6NTpD2o6qS12nE6sjc0UZ5x2imOCv7BGhdGkUqoFO4Ko6NJkSqJSrBpQpgnB/Nhs\n",
       "sglw+sd5rvcmbnbvJrt7f5z3a+ZO9v7M2ezm8zzP93yf80RKCUlSOea0egCSpNll8EtSYQx+SSqM\n",
       "wS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8\n",
       "klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9J\n",
       "hTH4JakwBr8kFcbgl6TCGPySVJi5rR6A1CoRMQgMVXf3ppT2t3I80mxxj19FiYh5EbFyUcS2Xtiz\n",
       "GLYvhu29sGdRxLaIWBkRva0epzSTIqXU6jFIs2JOxIoz4QOXQqyB+ddSP+Q9BmwFNsDBbZCOwOqn\n",
       "U9rcutFKM8fgVxH6Im4YgHV3Qv+ySV47DFwNoyNw4+GUbpqN8UmzyeBX15sTsWIINg1D/3lNvmcX\n",
       "sAxG98L17vmr2xj86moRMa8Pdt8LCy6b4nuHgSvhwGFYnFI6OhPjk1rByV11u+WXQEwU+p8AXgAM\n",
       "AmcDy4EfAsuApRDVQxpHRAxGxPnVbbDV41FzDH51tYWwdg3Mn+g1VwBfAvYD3wf6gT+rnlsD8xfC\n",
       "2pkdZWexM6rzGfzqWhExOAoXXtfw2Hrg2cAC8l7+3cBPA8+snk/AGcC51f3rgFG4yL3ZbE7Eij7Y\n",
       "/RK4eRMsPQQ9u2FgNwyMQM9GWHo53NwHe+ZErGj1eDU+g1/dbGgQxmotm98G/hH4BnAA+DzwnOq5\n",
       "e4GF5A3CLvIGAqAnP3YUOGuWxty2+iJuGIJN98CCL8P813D8GaA95JrYfTD/HlgwBJv6Im5ozWg1\n",
       "EYNfxTgDGAN2kPv2zwPOr557KbAPeJQcYH/eigG2sTkRKwZg3XAT7bCQ50eGoX8A1rnn337s6lHX\n",
       "iojBXtgzAj091WP/AvwTOfyvAv6Belmn5mvA1cCPyBuIATh2NHf2FLmkQzOdUR8C3kyeH6n5HPAM\n",
       "7IxqR+7xq2ullPb3w86tDY+tBO4hT+IG48/aHqMeYFvy1ztKDf3KpJ1RkCfJDzbcXoadUe3K4FdX\n",
       "2wfrN+Qc4iHyZO4YMA84k1z++Ti5rg95g/BXwGur++sh7YOBiFgdEQtmdfBt4sTOqPEmyCFPjI/H\n",
       "zqj2Y/Cr292+DdID5MB/B7CYXN55HHgPsJO8tzoA/CJwOfBe8glcD+aNxtvJ1Z9dEXFLRFwRETHb\n",
       "30grnNgZdbIJ8gD+i/xv+3xgHfBU9R47o9qPNX51velasiEilgC/Ry5nJ2Aj8JGU0u4ZGXgbiIjz\n",
       "F8P23Xm7yMPkjeTHyaWc2tzJd8l7kT8DPAisAH4X+Ivq+cVw6HG4OKX03dkcv8Zn8KsI07lIW7W3\n",
       "fwWwCng1cBfwQeDzKaWnTnx9u4uIOeRu1rOr2+KGr597NrxpT0PnZjMT5JuBvycfGYDB324MfhWj\n",
       "tizzJdWyzNdx/LLMW8jLMm+fwrLMVfni9eSjgHOBW4BNKaXvzcx3Mel4gjw3PV6In+yxs4ARcvVr\n",
       "T/Vn7TbSC389Amf0HP9XcRB4C/nf8CMnPLeZernMzqj2Y/CrKNVSAssXwtpRuKg6OYsD0NsPO/bl\n",
       "ucvbT6X1MCKWkjcAbyCXvD8IfCalNHYa4+0hXyVsvMA+WahDDvATQ3y8YN8DPJFSOnayMSyK2LYR\n",
       "li4nT5A/Sj7cCeCt5JrXCuBSYAnwLeB1wG8B7wRuA1bBth+ldOmp/jtoehn8Kla1t147I/eJ6dob\n",
       "jYgzgdeQNwJLgVvJ8wE7qK8F12yIPwN4gvEDe9xQTymNTsf30fD9rLwcbr4P5v83ub71TXJ9/wrg\n",
       "ZnK556Pkw4Yl5Pr+O8ldU5fDwa/C6pTSJ6ZzXDp1Br90iiKiVlKZKMSfDTyXvIEJ4Ah58c/G4J5o\n",
       "z3x/SunpWfumxuHS1t3H4Jf4iZLKRHvijfeDyffEa/d/RD6f6feBXwJuJx8FfCV1wH9CL2bTXQz+\n",
       "cVQlgKHq7l4npDpLNcHZ2KXSTIgPkEsqzYR47TZ6KqEdEedQbwt9inpb6J5T+oZniZev7B4GfyUi\n",
       "5lGf9LtwMJ/vw36Y1w87q0m/2zxcnX0NJZVmO1WGgENMLcT3zXZJpdpAvZS8AXg18AXyRuAL7doW\n",
       "OhOdUZp9Bj/1X+ZLq1/mazn+l3kr+Zd5m7/Mpy0i5nJ8SaWZdsMzmLgmfuL9vZ22ga6OMleSNwJL\n",
       "yG2ht7SqLXQiM9kZpdlRfPB7+Hrqqj3WWpdKs50q88n17mZDfA+nWFLpVCe0hT5Abgv97Om0hc6U\n",
       "meqM0swqOvidsDpeRPQxtRAfAg4zeWdK4/1ZL6l0qurnUWsLvZiqLTSl9GBLB6aOV2zwN9Oi9mHg\n",
       "/cD/kFcifAPwt8A22r9FrSqpnMXUesbnMrUTfzqupNKpIuK5wJuq2yPko4DNKaWDLR2YOlLJwf/j\n",
       "k1JO9pp/Ju9m/Tywm7zK4OvI68vO5kkpVUllAc2FeO3+AnJJpekTf4BDJZVUOlG1Qb+afB7Vy8kn\n",
       "xm4EvurPTs0qNvgbT0OHPBv1fvJSs88iL0L1ihPecxPwRXLnwumchl6d2dnMpGbt/hD5xJ+phPi+\n",
       "du0M0fSo2kLfSC4FHSNvAD7a7m2har0ig792Sb5D0DOXvMb4K4H7gXPIdfwnqV+PtebVwIXkck/D\n",
       "wlNLyGevT6VnvIfmT/ypdam03cSe2kN1RHgl+SjgOvIy+RuB/3Djr/GUGvxNrTHeaBPwLnJ9v9bC\n",
       "cDawN5+A09il0kynyoiH5ZoJEbGQelvoYuptod9v6cDUVgz+ykRrjH+GvArhXcBFDZ9TrTG+NKX0\n",
       "ndkZudS8iLiUvAFYCXydfBSwxaNHlRr8g72wZwR6Jltj/E7yufV3AC9ueJ1rjKtTVG2hy8kbgRcB\n",
       "HyO3he5o6cDUMkVeczeltL8fdm6t7p/sItx3A79NXk3rxSd8xhagH3YY+mp3KaXDKaVbU0qvAF5C\n",
       "bhT4QkTcFxFvjoiBST5CXabI4AfYB+s35B38cS/C/bfkC0YfBH6N3PM5H7imev8GOFidmi51jJTS\n",
       "wymlvwTOI19r/lrgkYj4YET8QikXkS9dkaUecI1xqSYizqXeFjpGvS308ZYOTDOm2D3+lNLYEVh9\n",
       "FYzumsL7dpHX6zmST94y9NXxUkqPpZT+DrgAeBv5KooPR8TmiPjV6mLs6iLF7vHXuEib9JOqttA3\n",
       "kI8ChsgdzbeklB5p6cA0LYoPfnCNcWkiJ7SF3k+9LdQj3g5l8FdcY1yaWNUW+lryRuBC8vXVN6aU\n",
       "vtnSgWnKDP5xuMa4NLGIeB5wPfkawt8lHwV8MqU00spxqTkGv6RTVq0W+uvko4CXAZ8ibwTud1mS\n",
       "9mXwS5oWDW2hq8gX6NkIfMy20PZj8EuaVtVJYC8nHwVcS175ZCNwl1dfaw8Gv6QZ09AWugpYRH21\n",
       "UNtCW8jglzQrIuIy8lHA64GvkS8f+a92ys0+g1/SrIqIfuptoS8kL4S7MaX0rZYOrCAGv6SWiYif\n",
       "pd4W+h3yUcCnUkqHWjmubmfwS2q5iOih3hb6UuptoV+3LXT6GfyS2kpEPIv6aqGj5KOAW1NKe1s6\n",
       "sC5i8EtqS9WqoLW20FcB/0Y+CrjbttDTY/BLansRsYjcFvoHwCB5A/ChlNKjLR1YhzL4JXWM6uSw\n",
       "xrbQr5A3AltTSsdaObZOYvBL6kgNbaGrgOdTbwv9dksH1gEMfkkdLyIuILeFvhF4mDwh/GnbQsdn\n",
       "8EvqGg1toauAK4BPkjcCw7aF1hn8krpSRPwU9bbQEeptoU+0dGBtwOCX1NUa2kJXAdcAd5A3Av85\n",
       "XW2h1cWbhqq7e9v94k0Gv6RiRMRZ1FcLXUC9LfQHp/BZ86hfrvXCQRgD2A/z+mFndbnW29pxETqD\n",
       "X1JxGtpCVwErgPvIRwGfa6YtdE7EijPhA5dCrIH51wJzq+eOAVuBDXBwG6QjsPrplDbPzHdyagx+\n",
       "SUWr2kJ/k7wRuAD4MLkt9KHxXt8XccMArLsT+pdN8tnDwNUwOgI3Hk7ppmkd+Gkw+CWpEhHPp94W\n",
       "+hD1ttBRyHv6Q7BpGPrPa/IzdwHLYHQvXN8ue/4GvySdoGoLvYZ8FPASYDPw4T7493thwWUned+D\n",
       "wNuBB4C9QG3meBi4Eg4chsXtUPOf0+oBSFK7SSkdSyl9JqX0KuBi4FHgsxfDwMlCH6CXvI7ExhMe\n",
       "XwYshQCWz8iAp8jgl6QJpJR+kFJ690J4bG1DZq4Hnk1uDXoBcDd5guBNwIXjfM4amL8Q1s7GmCdj\n",
       "qUeSJhERg72w5xD0zAW+DbwSuB84h1zHfxI4v3r9w+SNQONJAseAATh2NJd7Wtrn7x6/JE1uaBDG\n",
       "ai2bZ5Cb9neQA/086qF/Mj3AAjgKnDVjo2ySwS9JU/Q84H3Au4AlwErgsVYOaIoMfkma3N79MK/x\n",
       "zK6VwD3A98mztpMV748BB/L8b8vXCjL4JWkSKaX9/bBza3X/IfJk7hgwDziTXP4BOEKu51A9P1Z9\n",
       "vQXohx2tru+DwS9JTdkH6zfAQchh/g5gMXAu8DjwHuB7QD/wIvJRQB/wwur9G+BgtX5Py9nVI0lN\n",
       "iIh5fbB7ohO4TsYTuCSpA6WUxo7A6qtgdNcU3reLvF7PEVjdDqEPBr8kNe3plDaPwI3LYHS4idcP\n",
       "k9fpGYEb22WdHrDUI0lTVluW+ZJqWebrOH5Z5i3kmv52l2WWpO4REb3UL8RyUXVyFgegtx92VBO5\n",
       "t7dLeaeRwS9Jp6m69GLtjNwn2qFlcyIGvyQVxsldSSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiD\n",
       "X5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfgl\n",
       "qTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IK\n",
       "Y/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAG\n",
       "vyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IKY/BL\n",
       "UmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfglqTAGvyQV\n",
       "xuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVJi5rR5AO4qIQWCo\n",
       "urs3pbS/leORpOnkHn8lIuZFxMpFEdt6Yc9i2L4YtvfCnkUR2yJiZUT0tnqcknS6IqXU6jG03JyI\n",
       "FWfCBy6FWAPzr6V+KHQM2ApsgIPbIB2B1U+ntLl1o5Wk01N88PdF3DAA6+6E/mWTvHYYuBpGR+DG\n",
       "wyndNBvjk9R63Vb+LTr450SsGIJNw9B/XpPv2QUsg9G9cL17/lL3ioh5wPKFsHYULhyEMYD9MK8f\n",
       "du6D9cBtKaWjrR3p1BUb/BExrw923wsLLpvktb8MfBF4kjwpMgxcCQcOw+JO/KFLmli3l39Lntxd\n",
       "fgnEZKF/Kznwo+GxZcDS/NDymRqcpNboi7hhCDbdAwu+DPNfw/Htjz3k//j3wfx7YMEQbOqLuKE1\n",
       "oz01xQb/Qli7BuZDPl57NrAAeAFwd/Wa/cDfAO8FTjwuWgPzF8LaWRqupFkwJ2LFAKwbbmLOD/JO\n",
       "4DD0D8C6ORErZnp806XIUk9EDPbCnkPQ8x3glcD9wDnkGv6TwPnA24ALgN+o7tdKPZAP9wbg2NFc\n",
       "7unoiR5JzZV//xf4Y+BLwDzgevKOY6eVf0vd4x8ahLG5wBnkGZsd5DA/jxzy3wC+AvzRST6gB1gA\n",
       "R4GzZn64kmbBhOXfo+SdxF8B/g/4AfA71XOdVv4tNfh/7HnA+4B3AUuAlcAPgT+sHm/8Byrv2Egq\n",
       "x2Tl3w9Vj/0p0Af0Ahc3vL+Tyr9Fl3pGoKen4fGDwFuqP+8Anlk9/hTwOHnD8GngCiz1SN2kmfLv\n",
       "u8n/7x8Hvg68CHh/9Sd0ViYUucefUtrfDzu3Ag+Rt+Zj5JrdmcDZwGPA9up2R/W+B4Cfq77eAvTD\n",
       "jnb/AUtqyqTl30eATwB/Qs6Ha8jzf8eqD+ik8m+RwQ+wD9ZvgINjwDuAxcC55K35e8h7+7Xb2eTi\n",
       "3RLyDxdyD291AoekLjJe+fcxoB+4EriK3N65BtgLfKslozw9RZZ6YGoncJ2o02bwJU1ssvLvXOA5\n",
       "wJeBu6rnErAIuIdc67fU0wFSSmNHYPVVMLprCu/bRV6v5wisNvSl7jBZ+fcMcgfPV8nB/xT5qGAx\n",
       "8MLqMzqp/Fts8AM8ndLmEbhxGYwON/H6YfI6PSNwY6edoi1pYpOVfy8APga8lVzE30oO+9pZvZ1U\n",
       "/i221NOoti7HJdW6HNdx/LocW8g/1O0dui6HpMmVVP41+CvVRVZqK/FdVM3OcwB6+2FHtSW/vRN+\n",
       "qJJOTSkr9hr846jW3q61ZD3RCTU7SdOjhGt0GPySdIJuL/8a/JI0jm4u/xr8kjSJbiv/GvySVJii\n",
       "+/glqUQGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiD\n",
       "X5IKY/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXG4Jekwhj8klQYg1+SCmPwS1JhDH5JKozBL0mFMfgl\n",
       "qTAGvyQVxuCXpMIY/JJUGINfkgpj8EtSYQx+SSqMwS9JhTH4JakwBr8kFcbgl6TCGPySVBiDX5IK\n",
       "Y/BLUmEMfkkqjMEvSYUx+CWpMAa/JBXm/wHSTpgb0ym1cAAAAABJRU5ErkJggg==\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x106a83090>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Here we want our sequences to align from beginning to end (because they're toy examples)\n",
    "# so we'll penalize terminal gaps.\n",
    "from functools import partial \n",
    "global_pairwise_align_nucleotide = partial(global_pairwise_align_nucleotide, penalize_terminal_gaps=True)\n",
    "\n",
    "clusters, num_alignments = furthest_neighbor_cluster([s1, s2, s3, s4, s5, s6], 0.70, aligner=global_pairwise_align_nucleotide)\n",
    "print(num_alignments)\n",
    "show_clusters(clusters, plot_labels=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's apply that function to our sequence collection. This function will return the number of pairwise alignments that were performed (it'll be clear why we're interested in that soon), and the OTUs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU 1: ['995155', '4379525', '298988', '6749']\n",
      "OTU 2: ['877884', '243579', '82320', '327239', '694774', '730246', '4351185']\n",
      "OTU 3: ['731639', '819699', '605566']\n",
      "OTU 4: ['796969', '329302', '4433048', '4295874', '289174', '2931208', '251465', '164915']\n",
      "OTU 5: ['694282', '181529', '277215']\n",
      "OTU 6: ['572676', '4366868']\n",
      "OTU 7: ['275123', '181568', '310877', '182729', '4479562', '571031', '303919', '314337', '331117', '276257']\n",
      "OTU 8: ['566377', '181960', '167502']\n",
      "OTU 9: ['551423', '322469', '1118300', '4417725', '4431461', '239717', '815314']\n",
      "OTU 10: ['513370', '2532911', '730661', '4454871', '301867', '146275', '239868', '272591', '366311', '319497', '3427116', '885776']\n",
      "OTU 11: ['4417417', '358664']\n",
      "OTU 12: ['777102', '305957']\n",
      "OTU 13: ['298922']\n",
      "OTU 14: ['278960']\n",
      "OTU 15: ['261892']\n",
      "OTU 16: ['257001', '248801', '198774']\n",
      "OTU 17: ['250445']\n",
      "OTU 18: ['239787', '992327', '729865']\n",
      "OTU 19: ['222734', '20793']\n",
      "OTU 20: ['723052', '144260']\n",
      "OTU 21: ['107418']\n",
      "OTU 22: ['3984455', '104164']\n",
      "OTU 23: ['4475231', '83758']\n",
      "OTU 24: ['2338105']\n",
      "OTU 25: ['2660423']\n",
      "OTU 26: ['1082997']\n",
      "OTU 27: ['3053116']\n",
      "OTU 28: ['3625242']\n",
      "OTU 29: ['4301068']\n",
      "OTU 30: ['4334796']\n",
      "OTU 31: ['4354370']\n",
      "OTU 32: ['4401184']\n",
      "OTU 33: ['4422290']\n",
      "OTU 34: ['4423335', '4439805']\n",
      "OTU 35: ['4346018', '157389']\n",
      "OTU 36: ['100246', '108439']\n",
      "OTU 37: ['742091']\n"
     ]
    }
   ],
   "source": [
    "clusters, num_alignments = furthest_neighbor_cluster(seqs_16s, 0.70)\n",
    "show_clusters(clusters, plot_clusters=False, plot_labels=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's define a function that will be useful for exploring different clustering algorithms:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from time import time\n",
    "\n",
    "def evaluate_cluster_fn(cluster_fn, seqs, similarity_threshold, display=True):\n",
    "    start_time = time()\n",
    "    clusters, num_alignments = cluster_fn(seqs, similarity_threshold)\n",
    "    end_time = time()\n",
    "    run_time = end_time - start_time\n",
    "    num_clusters = len(clusters)\n",
    "    if display:\n",
    "        print(\"Number of alignments performed: %d\" % num_alignments)\n",
    "        print(\"Runtime: %1.3fs\" % run_time)\n",
    "        print(\"Number of clusters: %d\" % num_clusters)\n",
    "        print(\"Clusters:\")\n",
    "        show_clusters(clusters, plot_clusters=False, plot_labels=False)\n",
    "    return num_alignments, run_time, num_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's apply that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of alignments performed: 1458\n",
      "Runtime: 6.188s\n",
      "Number of clusters: 37\n",
      "Clusters:\n",
      "OTU 1: ['995155', '4379525', '298988', '6749']\n",
      "OTU 2: ['877884', '243579', '82320', '327239', '694774', '730246', '4351185']\n",
      "OTU 3: ['731639', '819699', '605566']\n",
      "OTU 4: ['796969', '329302', '4433048', '4295874', '289174', '2931208', '251465', '164915']\n",
      "OTU 5: ['694282', '181529', '277215']\n",
      "OTU 6: ['572676', '4366868']\n",
      "OTU 7: ['275123', '181568', '310877', '182729', '4479562', '571031', '303919', '314337', '331117', '276257']\n",
      "OTU 8: ['566377', '181960', '167502']\n",
      "OTU 9: ['551423', '322469', '1118300', '4417725', '4431461', '239717', '815314']\n",
      "OTU 10: ['513370', '2532911', '730661', '4454871', '301867', '146275', '239868', '272591', '366311', '319497', '3427116', '885776']\n",
      "OTU 11: ['4417417', '358664']\n",
      "OTU 12: ['777102', '305957']\n",
      "OTU 13: ['298922']\n",
      "OTU 14: ['278960']\n",
      "OTU 15: ['261892']\n",
      "OTU 16: ['257001', '248801', '198774']\n",
      "OTU 17: ['250445']\n",
      "OTU 18: ['239787', '992327', '729865']\n",
      "OTU 19: ['222734', '20793']\n",
      "OTU 20: ['723052', '144260']\n",
      "OTU 21: ['107418']\n",
      "OTU 22: ['3984455', '104164']\n",
      "OTU 23: ['4475231', '83758']\n",
      "OTU 24: ['2338105']\n",
      "OTU 25: ['2660423']\n",
      "OTU 26: ['1082997']\n",
      "OTU 27: ['3053116']\n",
      "OTU 28: ['3625242']\n",
      "OTU 29: ['4301068']\n",
      "OTU 30: ['4334796']\n",
      "OTU 31: ['4354370']\n",
      "OTU 32: ['4401184']\n",
      "OTU 33: ['4422290']\n",
      "OTU 34: ['4423335', '4439805']\n",
      "OTU 35: ['4346018', '157389']\n",
      "OTU 36: ['100246', '108439']\n",
      "OTU 37: ['742091']\n"
     ]
    }
   ],
   "source": [
    "r = evaluate_cluster_fn(furthest_neighbor_cluster, seqs_16s, 0.70)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Nearest neighbor clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try a variant on this algorithm. How would things change if **instead of requiring that a sequence be within the similarity treshold of all sequences in an OTU, we only required that it be within the similarity threshold of one sequence in that OTU**? This is referred to as **nearest neighbor** clustering, because cluster membership is defined by the percent similarity to the most similar (or *nearest*) \"neighbor\" in the cluster. \n",
    "\n",
    "Let's look at the process for six sequences again.\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.1.png\">\n",
    "\n",
    "Our first sequence, $S1$, will again define a new OTU. We'll call that OTU $OTU1$ and start building our mapping of OTUs to the sequences they contain:\n",
    "\n",
    "$OTU1$: $S1$ \n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.2.png\">\n",
    "\n",
    "Our second sequence, $S2$, still falls outside of the similarity threshold to $S1$, so will also define a new OTU. We'll call that OTU $OTU2$.\n",
    "\n",
    "$OTU1$: $S1$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.3.png\">\n",
    "\n",
    "Now imagine that our third sequence, $S3$ falls within the range of $OTU1$. We'd cluster $S3$ into $OTU1$ with $S1$. We now have three sequences clustered into two OTUs. So far, things are looking the same as before, except notice how our OTU definition (grey shading) is now different. Because any sequence within the similarity threshold of *any* of sequence in the OTU will fall into this OTU, the shading now covers the area covered by either of our sequences, rather than the area covered by both of our sequences (in set theory terminology, it is the *union* now, where previously it was the *intersection*). \n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.4.png\">\n",
    "\n",
    "Now let's cluster a fourth sequence, $S4$. We find that this falls outside the range of $OTU1$, and (just barely) outside the range of $OTU2$. So, we'd create a new OTU, $OTU3$, containing $S4$.\n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.5.png\">\n",
    "\n",
    "Next, let's cluster our fifth sequence, $S5$. We find that this falls outside the range of $OTU1$, and inside the range of both $OTU2$ and $OTU3$. As with furthest neighbor, we have a choice of how to handle this. We'll again assign it to the cluster to which it is the most similar. Our mapping of OTUs to sequences would look like:\n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$ $S5$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/nearest-neighbor.6.png\">\n",
    "\n",
    "Finally, let's cluster our last sequence, $S6$. Remember that $S6$ falls within the similarity range of $S1$, but outside of the similarity range of $S3$. In furthest neighbor, this meant that it was assigned to a new OTU, but with nearest neighbor it meets the inclusion criteria for $OTU1$. So, our final mapping of OTUs to sequences would look like:\n",
    "\n",
    "$OTU1$: $S1$ $S3$ $S6$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$ $S5$\n",
    "\n",
    "<hr>\n",
    "\n",
    "**One feature that becomes obvious here is the order dependence of these methods.** If sequences are provided in different order across different clustering runs, the cluster definitions will change. For example, how would the results differ if the sequences were processed in this order: $S1$, $S3$, $S4$, $S5$, $S6$, $S2$?\n",
    "\n",
    "<hr>\n",
    "\n",
    "Let's implement nearest neighbor clustering and explore its properties."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def nearest_neighbor_cluster(seqs, similarity_threshold):\n",
    "    clusters = []\n",
    "    num_alignments = 0\n",
    "    for query_seq in seqs:\n",
    "        best_percent_similarity = 0.0\n",
    "        cluster_index = None\n",
    "        for i, cluster in enumerate(clusters):\n",
    "            for cluster_seq_identifier in cluster:\n",
    "                current_cluster_representative_seq = seqs[cluster_seq_identifier]\n",
    "                aln = local_pairwise_align_ssw(str(query_seq), str(current_cluster_representative_seq))\n",
    "                num_alignments += 1\n",
    "                percent_similarity = aln[0].fraction_same(aln[1])\n",
    "                if percent_similarity >= similarity_threshold:\n",
    "                    if percent_similarity > best_percent_similarity:\n",
    "                        cluster_index = i\n",
    "                        best_percent_similarity = percent_similarity\n",
    "                        break\n",
    "        if cluster_index is None:\n",
    "            clusters.append([query_seq.id])\n",
    "        else:\n",
    "            clusters[cluster_index].append(query_seq.id)\n",
    "    return clusters, num_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = evaluate_cluster_fn(nearest_neighbor_cluster, seqs_16s, 0.70)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You'll notice that both the runtime and the number of alignments performed here are different. Most of the runtime is spent aligning, so runtime and number of alignments should be strongly correlated.\n",
    "\n",
    "There was another affect here though: we have a different number of OTUs. Is this result better or worse? There is not a definitive answer to that question: it really depends on the application, so what we'd ultimately want to know is how does that affect our ability to interpret the data. **Remember: OTU clustering is a necessary evil to deal with the massive amounts of data that we have. We don't necessary care about things like how many OTUs a method gives us, but rather how the clustering process helps or hurts us answer the biological questions driving the analysis.** We'll explore this concept more in later chapters, but it is an important one that algorithm developers sometimes lose track of."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Centroid clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, given that the number of alignments performed is correlated with runtime, are there ways that we can reduce the number of alignments that are computed by a clustering algorithm? One approach for that is generally referred to as **centroid clustering**. Here, we can say that **a sequence is assigned to an OTU if it is within the similarity threshold of the first sequence in that OTU**. The first sequence in that cluster then becomes the *cluster centroid*: cluster membership is defined by similarity to that one particular sequence, which effectively sits at the \"center\" of that OTU.\n",
    "\n",
    "Let's look at the process for our six sequences.\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.1.png\">\n",
    "\n",
    "Our first sequence, $S1$, will again define a new OTU. We'll call that OTU $OTU1$ and start building our mapping of OTUs to the sequences they contain:\n",
    "\n",
    "$OTU1$: $S1$ \n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.2.png\">\n",
    "\n",
    "Our second sequence, $S2$, still falls outside of the similarity threshold to $S1$, so will also define a new OTU. We'll call that OTU $OTU2$.\n",
    "\n",
    "$OTU1$: $S1$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.3.png\">\n",
    "\n",
    "Recall that $S3$ falls within the range of $OTU1$. We'd cluster $S3$ into $OTU1$ with $S1$, and now have three sequences clustered into two OTUs. Again, our sequence to OTU mapping looks the same as before at this stage, but our OTU definition (grey shading) is again different. Because a sequence must be within the similarity threshold of the first sequence added to the OTU (the *centroid*), the shading doesn't change. In other words, the definition of the OTU is fixed: additional sequences don't change the shaded area.\n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.4.png\">\n",
    "\n",
    "Now let's cluster a fourth sequence, $S4$. We find that this falls outside the range of $OTU1$, and (just barely) outside the range of $OTU2$. So, we'd create a new OTU, $OTU3$, containing $S4$.\n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.5.png\">\n",
    "\n",
    "Next, let's cluster our fifth sequence, $S5$. We find that this falls outside the range of $OTU1$, and inside the range of both $OTU2$ and $OTU3$. As with furthest neighbor, we have a choice of how to handle this. We'll again assign it to the cluster to which it is the most similar. Our mapping of OTUs to sequences would look like:\n",
    "\n",
    "$OTU1$: $S1$ $S3$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$ $S5$\n",
    "\n",
    "<hr>\n",
    "\n",
    "<img src=\"files/images/centroid-cluster.6.png\">\n",
    "\n",
    "Finally, let's cluster our last sequence, $S6$. Remember that $S6$ falls within the similarity range of $S1$, but outside of the similarity range of $S3$. In furthest neighbor, this meant that it was assigned to a new OTU; in nearest neighbor, it was assigned to $OTU1$, and that is what happens here (but how would this differ if $S3$ was the centroid of $OTU1$, rather than $S1$?). Our final mapping of OTUs to sequences would look like:\n",
    "\n",
    "$OTU1$: $S1$ $S3$ $S6$\n",
    "\n",
    "$OTU2$: $S2$\n",
    "\n",
    "$OTU3$: $S4$ $S5$\n",
    "\n",
    "<hr>\n",
    "\n",
    "Let's implment this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def centroid_cluster(seqs, similarity_threshold):\n",
    "    clusters = []\n",
    "    num_alignments = 0\n",
    "    for query_seq in seqs:\n",
    "        best_percent_similarity = 0.0\n",
    "        cluster_index = None\n",
    "        for i, cluster in enumerate(clusters):\n",
    "            current_cluster_representative_seq = seqs[cluster[0]]\n",
    "            aln = local_pairwise_align_ssw(str(query_seq), str(current_cluster_representative_seq))\n",
    "            num_alignments += 1\n",
    "            percent_similarity = aln[0].fraction_same(aln[1])\n",
    "            if percent_similarity >= similarity_threshold:\n",
    "                if percent_similarity > best_percent_similarity:\n",
    "                    cluster_index = i\n",
    "                    best_percent_similarity = percent_similarity\n",
    "        if cluster_index is None:\n",
    "            clusters.append([query_seq.id])\n",
    "        else:\n",
    "            clusters[cluster_index].append(query_seq.id)\n",
    "    return clusters, num_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = evaluate_cluster_fn(centroid_cluster, seqs_16s, 0.70)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've now reduced the number of alignments and the runtime. What was the effect on the results?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Three different definitions of OTUs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"float: right; margin-left: 30px;\"><img style=\"float: right; margin-left: 30px;\" src=\"files/images/cluster-types.png\" align=right></div>\n",
    "\n",
    "In these three algorithms, we've looked at three different ways of defining a cluster. **This figure illustrates the differences in each definition, where the solid lines indicate where a sequence must fall to be considered part of a cluster.**\n",
    "\n",
    "In nearest neighbor, where membership in a cluster is defined by a query sequence's distance to the most similar sequence already in the cluster (it's nearest neighbor), the size of the cluster can grow with additional sequences. This can have some undesired effects: for example, if we clustered our sequences in order of decreasing pairwise similarity, in the worst case we could end up with one single cluster containing all of our sequences.\n",
    "\n",
    "In furthest neighbor, where membership in a cluster is defined by a query sequence's distance to the most dissimilar sequence already in the cluster (it's furthest neighbor), the size of the cluster can shrink with additional sequences. This can also have some undesired effects, such as the size of a cluster becoming contstrained to the point where it is unlikely that new sequences will ever be added.\n",
    "\n",
    "In centroid distance, where membership in a cluster is defined by a query sequence's distance to the cluster's centroid sequence, the size of the cluster always remains the same, but the role that the first sequence added to a cluster plays becomes much more important. So, it's very important that the cluster centroids are well-chosen. One undesired effect of centroid distance cluster is that the cluster sizes are fixed, which may or may not always make biological sense (for example, if the marker gene evolves at a faster rate in some taxa than others, that can't be reflected in the cluster defintions. \n",
    "\n",
    "All of these methods have good features and bad features, and that in fact is a common feature of heuristics (if they were perfect, they wouldn't be heuristics after all...). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Greedy centroid clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While we're looking at heuristics, let's look at one more heuristic. Remember that in the multiple sequence alignment chapter we discussed computing *kmer distances* between sequences, which is an alignment-independent way of getting an idea of the similarity of a pair of sequences. Relative to pairwise alignment, this is very fast, so another strategy we can take is to compute all kmer distances between a query sequence and all cluster centroids first. Then, run our centroid clustering process, but only compute the pairwise alignment between a query sequence and the cluster centroid that has the smallest kmer distance to the query. The affect will be that rather than computing the pairwise alignment between a query and all cluster centroids, we only compute the alignment between the query sequence and the cluster centroid that we think is likely to be most similar to the query sequence, based on it having the smallest kmer distance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from iab.algorithms import kmer_distance\n",
    "from functools import partial\n",
    "\n",
    "# cache the kmer distances between all sequences\n",
    "# need to improve the code below so this is calculate on-the-fly only as necessary, but waiting\n",
    "# on faster kmer computation in skbio as this is actually currently slower than SSW alignment\n",
    "seqs_16s_5mer_distances = seqs_16s.distances(partial(kmer_distance, k=5, overlapping=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def greedy_centroid_cluster(seqs, similarity_threshold, kmer_distances, max_rejects=5):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    clusters = {}\n",
    "    num_alignments = 0\n",
    "\n",
    "    for query_seq in seqs:\n",
    "        # find the centroids with the smallest kmer distance to the query seq\n",
    "        current_kmer_distances = []\n",
    "        for cluster_centroid_id in clusters:\n",
    "            current_kmer_distances.append((kmer_distances[(query_seq.id, cluster_centroid_id)], cluster_centroid_id))\n",
    "        current_kmer_distances.sort()\n",
    "        \n",
    "        best_percent_similarity = 0.0\n",
    "        best_cluster_centroid_id = None\n",
    "        # align the query against that centroid and compute percent similarity\n",
    "        for _, cluster_centroid_id in current_kmer_distances[:max_rejects]:\n",
    "            aln = local_pairwise_align_ssw(str(query_seq), str(seqs.get_seq(cluster_centroid_id)))\n",
    "            num_alignments += 1\n",
    "            percent_similarity = aln[0].fraction_same(aln[1])\n",
    "\n",
    "            # and if it's within the similarity threshold, add the sequence to that cluster\n",
    "            if percent_similarity >= similarity_threshold and percent_similarity > best_percent_similarity:\n",
    "                best_percent_similarity = percent_similarity\n",
    "                best_cluster_centroid_id = cluster_centroid_id\n",
    "\n",
    "        if best_cluster_centroid_id is None:\n",
    "            clusters[query_seq.id] = [query_seq.id]\n",
    "        else:\n",
    "            clusters[best_cluster_centroid_id].append(query_seq.id)\n",
    "    \n",
    "    clusters = clusters.values()\n",
    "    return clusters, num_alignments\n",
    "\n",
    "greedy_centroid_cluster = partial(greedy_centroid_cluster, kmer_distances=seqs_16s_5mer_distances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = evaluate_cluster_fn(greedy_centroid_cluster, seqs_16s, 0.70)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do these results compare to the previous methods?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing properties of our clustering algorithms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We so far looked at these algorithms based on a single similarity threshold and a single sequence collection, but as we know from previous chapters it's important to know how features such as run time change with different inputs. Let's explore these algorithms in the context of changing sequence collection sizes and similarity thresholds. \n",
    "\n",
    "For the sake of runtime, I'm only looking at three settings for each. You may want to expand from there, but don't work with more than 5 or 7 different similarity thresholds or sequence collection sizes as the runtimes will increase rapidly. Note that every additional value in the ``similarity_threshold`` or ``sequence_collections`` list will add four clustering runs, since each method is run at each similarity threshold (in this next cell) or on each sequence collection (a few cells further down). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cluster_fns = [(\"furthest neighbor\", furthest_neighbor_cluster), \n",
    "               (\"nearest_neighbor\", nearest_neighbor_cluster), \n",
    "               (\"centroid\", centroid_cluster),\n",
    "               (\"greedy centroid\", greedy_centroid_cluster)]\n",
    "similarity_thresholds = [0.60, 0.70, 0.80, 0.90]\n",
    "sequence_collections = map(SequenceCollection,[seqs_16s[:25], seqs_16s[:50], seqs_16s[:75], seqs_16s[:100]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First let's look at different similarity thresholds by looping over our clustering functions and a few similarity thresholds, and finding the run time for each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "alignment_counts = []\n",
    "run_times = []\n",
    "cluster_counts = []\n",
    "for cluster_fn in cluster_fns:\n",
    "    current_alignment_counts = []\n",
    "    current_run_times = []\n",
    "    current_cluster_counts = []\n",
    "    for similarity_threshold in similarity_thresholds:\n",
    "        num_alignments, run_time, num_clusters = evaluate_cluster_fn(\n",
    "            cluster_fn[1], sequence_collections[0], similarity_threshold, display=False)\n",
    "        current_run_times.append(run_time)\n",
    "        current_alignment_counts.append(num_alignments)\n",
    "        current_cluster_counts.append(num_clusters)\n",
    "    alignment_counts.append(current_alignment_counts)\n",
    "    run_times.append(current_run_times)\n",
    "    cluster_counts.append(current_cluster_counts)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now plot these using matplotlib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for rt in run_times:\n",
    "    plt.plot(similarity_thresholds, rt)\n",
    "plt.legend([e[0] for e in cluster_fns], loc=\"upper left\")\n",
    "plt.xlabel(\"Similarity threshold\")\n",
    "plt.ylabel(\"Run time (s)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember that above I said that most of the time in each of these clustering algorithms is spent doing pairwise alignment. Let's plot the number of alignments comptued by each as well so I can prove that to you."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for na in alignment_counts:\n",
    "    plt.plot(similarity_thresholds, na)\n",
    "plt.legend([e[0] for e in cluster_fns], loc=\"upper left\")\n",
    "plt.xlabel(\"Similarity threshold\")\n",
    "plt.ylabel(\"Number of alignments computed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next let's look at run time as a function of the number of sequences to be clustered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "alignment_counts = []\n",
    "run_times = []\n",
    "cluster_counts = []\n",
    "for cluster_fn in cluster_fns:\n",
    "    current_alignment_counts = []\n",
    "    current_run_times = []\n",
    "    current_cluster_counts = []\n",
    "    for sequence_collection in sequence_collections:\n",
    "        num_alignments, run_time, num_clusters = evaluate_cluster_fn(cluster_fn[1], sequence_collection, 0.70, display=False)\n",
    "        current_run_times.append(run_time)\n",
    "        current_alignment_counts.append(num_alignments)\n",
    "        current_cluster_counts.append(num_clusters)\n",
    "    alignment_counts.append(current_alignment_counts)\n",
    "    run_times.append(current_run_times)\n",
    "    cluster_counts.append(current_cluster_counts) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for rt in run_times:\n",
    "    plt.plot([e.sequence_count() for e in sequence_collections], rt)\n",
    "plt.legend([e[0] for e in cluster_fns], loc=\"upper left\")\n",
    "plt.xlabel(\"Number of sequences\")\n",
    "plt.ylabel(\"Run time (s)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Which of these methods do you think will scale best** to continuosuly increasing numbers of sequences (e.g., as is currently the trend in microbiomics)?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reference-based clustering to assist with parallelization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**This section needs work...**\n",
    "\n",
    "**TODO**\n",
    "\n",
    "* Discuss issues with run time (maybe examples from EMP clustering run times)\n",
    "* Need for parallelization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Up until this point we have focused our discussion on *de novo* OTU clustering, meaning that sequences are clustered only against each other, with no external reference. This is a very widely applied protocol, and the primary function of popular bioinformatics tools such as [cdhit](http://bioinformatics.oxfordjournals.org/content/28/23/3150.long) and [uclust](http://bioinformatics.oxfordjournals.org/content/26/19/2460.long). Another category of OTU clustering protocols is also popular however: reference-based OTU clustering, where a external reference database of sequences is used to aid in cluster defintion. In this section we compare *de novo* clustering with two reference-based OTU clustering protocols, closed-reference and open-reference. \n",
    "\n",
    "De novo OTU clustering\n",
    "----------------------\n",
    "\n",
    "In a de novo OTU clustering process, as discussed above, reads are clustered against one another without any external reference sequence collection. A benefit of de novo OTU picking is that all reads are clustered. A drawback is that there is no existing support for running this in parallel in QIIME, so it can be too slow to apply to large datasets (e.g., more than 10 million reads). \n",
    "\n",
    "You **must** use de novo OTU picking if:\n",
    "\n",
    "*  You do not have a reference sequence collection to cluster against, for example because you're working with an infrequently used marker gene.\n",
    "\n",
    "You **cannot** use de novo OTU picking if:\n",
    "\n",
    "*  You are comparing non-overlapping loci, such as the V2 and the V4 regions of the 16S rRNA.\n",
    "*  You working with very large data sets, like a full HiSeq 2000 run. (Technically, you can use de novo OTU picking here, but you literally might wait a month or more for your results.)\n",
    "\n",
    "Pros:\n",
    "\n",
    "*  All reads are clustered\n",
    "\n",
    "Cons:\n",
    "\n",
    "*  Speed. Does not run in parallel.\n",
    "\n",
    "Closed-reference OTU clustering (also referred to as *read mapping*)\n",
    "--------------------------------------------------------------------\n",
    "\n",
    "In a closed-reference OTU clustering process, reads are clustered against a reference sequence collection and any reads which do not hit a sequence in the reference sequence collection are excluded from downstream analyses. \n",
    "\n",
    "You **must** use closed-reference OTU clustering if:\n",
    "\n",
    "*  You are comparing non-overlapping loci, such as the V2 and the V4 regions of the 16S rRNA. Your reference sequences must span both of the regions being sequenced.\n",
    "\n",
    "You **cannot** use closed-reference OTU clustering if:\n",
    "\n",
    "*  You do not have a reference sequence collection to cluster against, for example because you're working with an infrequently used marker gene.\n",
    "\n",
    "Pros:\n",
    "\n",
    "*  Speed. Closed-reference OTU clustering is fully parallelizable, so is useful for extremely large data sets.\n",
    "*  Better trees and taxonomy. Because all OTUs are already defined in your reference sequence collection, you may already have a tree and a taxonomy that you trust for those OTUs. You have the option of using those, or building a tree and taxonomy from your sequence data.\n",
    "\n",
    "Cons:\n",
    "\n",
    "*  Inability to detect novel diversity with respect to your reference sequence collection. Because reads that don't hit the reference sequence collection are discarded, your subsequent analyses can only focus on the diversity that you \"already know about\". Depending on how well-characterized the environment that you're working in is, you may end up throwing away a small fraction of your reads (e.g., discarding 1-10% of the reads is common for 16S-based human microbiome studies, where databases like Greengenes cover most of the organisms that are typically present) or a large fraction of your reads (e.g, discarding 50-80% of the reads has been observed for \"unusual\" environments like [the Guerrero Negro microbial mats](http://www.ncbi.nlm.nih.gov/pubmed/22832344)). \n",
    "\n",
    "Open-reference OTU clustering\n",
    "-----------------------------\n",
    "\n",
    "In an open-reference OTU clustering process, reads are clustered against a reference sequence collection and any reads which do not hit the reference sequence collection are subsequently clustered de novo.\n",
    "\n",
    "You **cannot** use open-reference OTU clustering if:\n",
    "\n",
    "*  You are comparing non-overlapping loci, such as the V2 and the V4 regions of the 16S rRNA.\n",
    "*  You do not have a reference sequence collection to cluster against, for example because you're working with an infrequently used marker gene.\n",
    "\n",
    "Pros:\n",
    "\n",
    "*  All reads are clustered.\n",
    "*  Speed. Open-reference OTU clustering can partially run in parallel.\n",
    "\n",
    "Cons:\n",
    "\n",
    "*  Speed. Some steps of this protocol do still run serially. For data sets with a lot of novel diversity with respect to the reference sequence collection, this can still take days to run.\n",
    "\n",
    "OTU clustering in QIIME\n",
    "-----------------------\n",
    "\n",
    "The [QIIME](www.qiime.org) software package implements these three OTU clustering methods, with some variations on the underlying clustering algorithms that are applied for each. The [QIIME OTU picking tutorial](http://qiime.org/tutorials/otu_picking.html), which this section of this chapter was derived from, covers how to access these different methods.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
